# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [2]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [3]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    # TODO: Implement Function
    source_id_text = [[source_vocab_to_int[word] for word in sentence.split()] for sentence in source_text.split('\n')]
    
    target_id_text = [[target_vocab_to_int[word] for word in sentence.split()] + [target_vocab_to_int['<EOS>']] for sentence in target_text.split('\n')]
    
    
    return source_id_text, target_id_text


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [4]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [5]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf
from tensorflow.python.layers.core import Dense

# Check TensorFlow Version
assert LooseVersion(tf.__version__) >= LooseVersion('1.1'), 'Please use TensorFlow version 1.1 or newer'
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.1.0


/home/aash/anaconda3/envs/tensorflow/lib/python3.5/site-packages/ipykernel_launcher.py:15: UserWarning: No GPU found. Please use a GPU to train your neural network.
  from ipykernel import kernelapp as app


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoder_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.
- Target sequence length placeholder named "target_sequence_length" with rank 1
- Max target sequence length tensor named "max_target_len" getting its value from applying tf.reduce_max on the target_sequence_length placeholder. Rank 0.
- Source sequence length placeholder named "source_sequence_length" with rank 1

Return the placeholders in the following the tuple (input, targets, learning rate, keep probability, target sequence length, max target sequence length, source sequence length)

In [7]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, learning rate, and lengths of source and target sequences.
    :return: Tuple (input, targets, learning rate, keep probability, target sequence length,
    max target sequence length, source sequence length)
    """
    # TODO: Implement Function
    input_int_text = tf.placeholder(tf.int32, [None, None], name ='input')
    targets_int_text = tf.placeholder(tf.int32, [None, None], name = 'targets')
    learning_rate = tf.placeholder(tf.float32, name = 'learning_rate')
    keep_probability = tf.placeholder(tf.float32, name = 'keep_prob')
    target_seq_length = tf.placeholder(tf.int32, (None,), name = 'target_sequence_length')
    max_target_seq_length = tf.reduce_max(target_seq_length)
    source_seq_length = tf.placeholder(tf.int32,(None,), name = 'source_sequence_length')
    
    
    return input_int_text, targets_int_text, learning_rate, keep_probability, target_seq_length,max_target_seq_length, source_seq_length


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

Tests Passed


### Process Decoder Input
Implement `process_decoder_input` by removing the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [8]:
def process_decoder_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for encoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    # TODO: Implement Function
    
    
    ending = tf.strided_slice(target_data, [0, 0], [batch_size, -1], [1, 1])
    processed_target = tf.concat([tf.fill([batch_size, 1], target_vocab_to_int['<GO>']), ending], 1)
    return processed_target
    

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_encoding_input(process_decoder_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer:
 * Embed the encoder input using [`tf.contrib.layers.embed_sequence`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence)
 * Construct a [stacked](https://github.com/tensorflow/tensorflow/blob/6947f65a374ebf29e74bb71e36fd82760056d82c/tensorflow/docs_src/tutorials/recurrent.md#stacking-multiple-lstms) [`tf.contrib.rnn.LSTMCell`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) wrapped in a [`tf.contrib.rnn.DropoutWrapper`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper)
 * Pass cell and embedded input to [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn)

In [9]:
from imp import reload
reload(tests)

def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob, 
                   source_sequence_length,source_vocab_size, 
                   encoding_embedding_size):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :param source_sequence_length: a list of the lengths of each sequence in the batch
    :param source_vocab_size: vocabulary size of source data
    :param encoding_embedding_size: embedding size of source data
    :return: tuple (RNN output, RNN state)
    """
    # TODO: Implement Function
    
    enc_embed_input = tf.contrib.layers.embed_sequence(rnn_inputs, source_vocab_size, encoding_embedding_size)
    
    def make_cell(rnn_size):
        enc_cell = tf.contrib.rnn.LSTMCell(rnn_size, initializer=tf.random_uniform_initializer(-0.1, 0.1, seed=2))
        return enc_cell
    
    enc_cell = tf.contrib.rnn.MultiRNNCell([make_cell(rnn_size) for _ in range(num_layers)])
    
    enc_cell = tf.contrib.rnn.DropoutWrapper(enc_cell, output_keep_prob=keep_prob)
                      
    enc_output, enc_state = tf.nn.dynamic_rnn(enc_cell, enc_embed_input, sequence_length=source_sequence_length, dtype=tf.float32)
                   
    return enc_output, enc_state


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create a training decoding layer:
* Create a [`tf.contrib.seq2seq.TrainingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/TrainingHelper) 
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [10]:

def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_summary_length, 
                         output_layer, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_summary_length: The length of the longest sequence in the batch
    :param output_layer: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing training logits and sample_id
    """
    # TODO: Implement Function
    
    helper = tf.contrib.seq2seq.TrainingHelper(dec_embed_input, target_sequence_length, time_major=False)
    decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell, helper, encoder_state, output_layer)
    
    final_outputs, _ = tf.contrib.seq2seq.dynamic_decode(decoder, impute_finished=True,
                                                         maximum_iterations=max_summary_length)
    

    
    return final_outputs



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference decoder:
* Create a [`tf.contrib.seq2seq.GreedyEmbeddingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/GreedyEmbeddingHelper)
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [11]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id,
                         end_of_sequence_id, max_target_sequence_length,
                         vocab_size, output_layer, batch_size, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param max_target_sequence_length: Maximum length of target sequences
    :param vocab_size: Size of decoder/target vocabulary
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_layer: Function to apply the output layer
    :param batch_size: Batch size
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing inference logits and sample_id
    """
    # TODO: Implement Function
    start_tokens = tf.tile(tf.constant([start_of_sequence_id], dtype=tf.int32), [batch_size], name='start_tokens')
    
    helper =  tf.contrib.seq2seq.GreedyEmbeddingHelper(dec_embeddings, start_tokens, end_of_sequence_id)
    
    decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell, helper, encoder_state, output_layer)
    
    finaloutputs, _  = tf.contrib.seq2seq.dynamic_decode(decoder,impute_finished=True,
                                                         maximum_iterations=max_target_sequence_length)
    return finaloutputs



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

* Embed the target sequences
* Construct the decoder LSTM cell (just like you constructed the encoder cell above)
* Create an output layer to map the outputs of the decoder to the elements of our vocabulary
* Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)` function to get the training logits.
* Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [12]:
def decoding_layer(dec_input, encoder_state,
                   target_sequence_length, max_target_sequence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, decoding_embedding_size):
    """
    Create decoding layer
    :param dec_input: Decoder input
    :param encoder_state: Encoder state
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_target_sequence_length: Maximum length of target sequences
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param target_vocab_size: Size of target vocabulary
    :param batch_size: The size of the batch
    :param keep_prob: Dropout keep probability
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    start_of_sequence_id = target_vocab_to_int['<GO>']
    end_of_sequence_id = target_vocab_to_int['<EOS>']

    dec_embeddings = tf.Variable(tf.random_uniform([target_vocab_size, decoding_embedding_size]))
    dec_embed_input = tf.nn.embedding_lookup(dec_embeddings, dec_input)
      
    #decoder cell
    def make_cell(rnn_size):
        dec_cell = tf.contrib.rnn.LSTMCell(rnn_size,
                                           initializer=tf.random_uniform_initializer(-0.1, 0.1, seed=2))
        return dec_cell

    dec_cell = tf.contrib.rnn.MultiRNNCell([make_cell(rnn_size) for _ in range(num_layers)])
       
    output_layer = Dense(target_vocab_size, kernel_initializer =
                         tf.truncated_normal_initializer(mean=0.0, stddev=0.1 ))
          
    with tf.variable_scope('decode'):       
       
       
        train_basic_dec_output = decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length,        
                                                      max_target_sequence_length, output_layer, keep_prob)
       
       
       
    with tf.variable_scope('decode', reuse=True):                              
        
        inf_basic_dec_output = decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id,
                                                       end_of_sequence_id, max_target_sequence_length, target_vocab_size,
                                                       output_layer, batch_size, keep_prob)
       
       
       
       
    return train_basic_dec_output, inf_basic_dec_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Apply embedding to the input data for the encoder.
- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size)`.
- Process target data using your `process_decoder_input(target_data, target_vocab_to_int, batch_size)` function.
- Apply embedding to the target data for the decoder.
- Decode the encoded input using your `decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)` function.

In [14]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size,
                  source_sequence_length, target_sequence_length,
                  max_target_sentence_length,
                  source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size,
                  rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param source_sequence_length: Sequence Lengths of source sequences in the batch
    :param target_sequence_length: Sequence Lengths of target sequences in the batch
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    
    inp_encode, enc_state = encoding_layer(input_data, rnn_size, num_layers, keep_prob,  source_sequence_length, 
                   source_vocab_size, enc_embedding_size)
    
    processed_input = process_decoder_input(target_data, target_vocab_to_int, batch_size)    

    training_logits, inference_logits = decoding_layer(processed_input, enc_state, target_sequence_length, 
                                       max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, 
                                       target_vocab_size, batch_size, keep_prob, dec_embedding_size)
    
    
    return training_logits, inference_logits


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability
- Set `display_step` to state how many steps between each debug output statement

In [15]:
# Number of Epochs
epochs = 10
# Batch Size
batch_size = 64
# RNN Size
rnn_size = 32
# Number of Layers
num_layers = 2
# Embedding Size
encoding_embedding_size = 64
decoding_embedding_size = 64
# Learning Rate
learning_rate = 0.005
# Dropout Keep Probability
keep_probability = 0.8
display_step = 4

### Build the Graph
Build the graph using the neural network you implemented.

In [16]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_target_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length = model_inputs()

    #sequence_length = tf.placeholder_with_default(max_target_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)

    train_logits, inference_logits = seq2seq_model(tf.reverse(input_data, [-1]),
                                                   targets,
                                                   keep_prob,
                                                   batch_size,
                                                   source_sequence_length,
                                                   target_sequence_length,
                                                   max_target_sequence_length,
                                                   len(source_vocab_to_int),
                                                   len(target_vocab_to_int),
                                                   encoding_embedding_size,
                                                   decoding_embedding_size,
                                                   rnn_size,
                                                   num_layers,
                                                   target_vocab_to_int)


    training_logits = tf.identity(train_logits.rnn_output, name='logits')
    inference_logits = tf.identity(inference_logits.sample_id, name='predictions')

    masks = tf.sequence_mask(target_sequence_length, max_target_sequence_length, dtype=tf.float32, name='masks')

    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            training_logits,
            targets,
            masks)

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)


Batch and pad the source and target sequences

In [17]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def pad_sentence_batch(sentence_batch, pad_int):
    """Pad sentences with <PAD> so that each sentence of a batch has the same length"""
    max_sentence = max([len(sentence) for sentence in sentence_batch])
    return [sentence + [pad_int] * (max_sentence - len(sentence)) for sentence in sentence_batch]


def get_batches(sources, targets, batch_size, source_pad_int, target_pad_int):
    """Batch targets, sources, and the lengths of their sentences together"""
    for batch_i in range(0, len(sources)//batch_size):
        start_i = batch_i * batch_size

        # Slice the right amount for the batch
        sources_batch = sources[start_i:start_i + batch_size]
        targets_batch = targets[start_i:start_i + batch_size]

        # Pad
        pad_sources_batch = np.array(pad_sentence_batch(sources_batch, source_pad_int))
        pad_targets_batch = np.array(pad_sentence_batch(targets_batch, target_pad_int))

        # Need the lengths for the _lengths parameters
        pad_targets_lengths = []
        for target in pad_targets_batch:
            pad_targets_lengths.append(len(target))

        pad_source_lengths = []
        for source in pad_sources_batch:
            pad_source_lengths.append(len(source))

        yield pad_sources_batch, pad_targets_batch, pad_source_lengths, pad_targets_lengths


### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [18]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1])],
            'constant')

    return np.mean(np.equal(target, logits))

# Split data to training and validation sets
train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]
valid_source = source_int_text[:batch_size]
valid_target = target_int_text[:batch_size]
(valid_sources_batch, valid_targets_batch, valid_sources_lengths, valid_targets_lengths ) = next(get_batches(valid_source,
                                                                                                             valid_target,
                                                                                                             batch_size,
                                                                                                             source_vocab_to_int['<PAD>'],
                                                                                                             target_vocab_to_int['<PAD>']))                                                                                                  
with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch, sources_lengths, targets_lengths) in enumerate(
                get_batches(train_source, train_target, batch_size,
                            source_vocab_to_int['<PAD>'],
                            target_vocab_to_int['<PAD>'])):

            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 target_sequence_length: targets_lengths,
                 source_sequence_length: sources_lengths,
                 keep_prob: keep_probability})


            if batch_i % display_step == 0 and batch_i > 0:


                batch_train_logits = sess.run(
                    inference_logits,
                    {input_data: source_batch,
                     source_sequence_length: sources_lengths,
                     target_sequence_length: targets_lengths,
                     keep_prob: 1.0})


                batch_valid_logits = sess.run(
                    inference_logits,
                    {input_data: valid_sources_batch,
                     source_sequence_length: valid_sources_lengths,
                     target_sequence_length: valid_targets_lengths,
                     keep_prob: 1.0})

                train_acc = get_accuracy(target_batch, batch_train_logits)

                valid_acc = get_accuracy(valid_targets_batch, batch_valid_logits)

                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.4f}, Validation Accuracy: {:>6.4f}, Loss: {:>6.4f}'
                      .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch    4/2154 - Train Accuracy: 0.2172, Validation Accuracy: 0.2841, Loss: 5.4041
Epoch   0 Batch    8/2154 - Train Accuracy: 0.1850, Validation Accuracy: 0.2841, Loss: 4.7235
Epoch   0 Batch   12/2154 - Train Accuracy: 0.2445, Validation Accuracy: 0.2841, Loss: 4.0614
Epoch   0 Batch   16/2154 - Train Accuracy: 0.2352, Validation Accuracy: 0.2997, Loss: 3.7437
Epoch   0 Batch   20/2154 - Train Accuracy: 0.2797, Validation Accuracy: 0.3189, Loss: 3.5451
Epoch   0 Batch   24/2154 - Train Accuracy: 0.2790, Validation Accuracy: 0.3161, Loss: 3.4912
Epoch   0 Batch   28/2154 - Train Accuracy: 0.3021, Validation Accuracy: 0.3189, Loss: 3.4300
Epoch   0 Batch   32/2154 - Train Accuracy: 0.2467, Validation Accuracy: 0.3182, Loss: 3.5174
Epoch   0 Batch   36/2154 - Train Accuracy: 0.2836, Validation Accuracy: 0.3182, Loss: 3.2928
Epoch   0 Batch   40/2154 - Train Accuracy: 0.2262, Validation Accuracy: 0.3189, Loss: 3.4147
Epoch   0 Batch   44/2154 - Train Accuracy: 0.2383, Validati

Epoch   0 Batch  356/2154 - Train Accuracy: 0.3766, Validation Accuracy: 0.4268, Loss: 1.4643
Epoch   0 Batch  360/2154 - Train Accuracy: 0.3709, Validation Accuracy: 0.4439, Loss: 1.4282
Epoch   0 Batch  364/2154 - Train Accuracy: 0.4062, Validation Accuracy: 0.4276, Loss: 1.3796
Epoch   0 Batch  368/2154 - Train Accuracy: 0.3898, Validation Accuracy: 0.4020, Loss: 1.4342
Epoch   0 Batch  372/2154 - Train Accuracy: 0.3413, Validation Accuracy: 0.4162, Loss: 1.4298
Epoch   0 Batch  376/2154 - Train Accuracy: 0.3734, Validation Accuracy: 0.4062, Loss: 1.3461
Epoch   0 Batch  380/2154 - Train Accuracy: 0.3461, Validation Accuracy: 0.4254, Loss: 1.3372
Epoch   0 Batch  384/2154 - Train Accuracy: 0.4112, Validation Accuracy: 0.4084, Loss: 1.2194
Epoch   0 Batch  388/2154 - Train Accuracy: 0.4375, Validation Accuracy: 0.3878, Loss: 1.2726
Epoch   0 Batch  392/2154 - Train Accuracy: 0.3773, Validation Accuracy: 0.3935, Loss: 1.3163
Epoch   0 Batch  396/2154 - Train Accuracy: 0.3561, Validati

Epoch   0 Batch  708/2154 - Train Accuracy: 0.4885, Validation Accuracy: 0.4993, Loss: 0.9323
Epoch   0 Batch  712/2154 - Train Accuracy: 0.4836, Validation Accuracy: 0.5078, Loss: 0.8438
Epoch   0 Batch  716/2154 - Train Accuracy: 0.4918, Validation Accuracy: 0.4972, Loss: 0.8087
Epoch   0 Batch  720/2154 - Train Accuracy: 0.5058, Validation Accuracy: 0.5128, Loss: 0.8613
Epoch   0 Batch  724/2154 - Train Accuracy: 0.5016, Validation Accuracy: 0.5014, Loss: 0.8711
Epoch   0 Batch  728/2154 - Train Accuracy: 0.5102, Validation Accuracy: 0.5213, Loss: 0.8348
Epoch   0 Batch  732/2154 - Train Accuracy: 0.4926, Validation Accuracy: 0.5142, Loss: 0.8461
Epoch   0 Batch  736/2154 - Train Accuracy: 0.5109, Validation Accuracy: 0.4979, Loss: 0.7466
Epoch   0 Batch  740/2154 - Train Accuracy: 0.5406, Validation Accuracy: 0.5071, Loss: 0.7918
Epoch   0 Batch  744/2154 - Train Accuracy: 0.4898, Validation Accuracy: 0.5000, Loss: 0.8012
Epoch   0 Batch  748/2154 - Train Accuracy: 0.5335, Validati

Epoch   0 Batch 1060/2154 - Train Accuracy: 0.5703, Validation Accuracy: 0.5639, Loss: 0.6374
Epoch   0 Batch 1064/2154 - Train Accuracy: 0.5633, Validation Accuracy: 0.5618, Loss: 0.6273
Epoch   0 Batch 1068/2154 - Train Accuracy: 0.5567, Validation Accuracy: 0.5888, Loss: 0.6658
Epoch   0 Batch 1072/2154 - Train Accuracy: 0.5687, Validation Accuracy: 0.5938, Loss: 0.6225
Epoch   0 Batch 1076/2154 - Train Accuracy: 0.5872, Validation Accuracy: 0.5916, Loss: 0.6848
Epoch   0 Batch 1080/2154 - Train Accuracy: 0.5672, Validation Accuracy: 0.6108, Loss: 0.6231
Epoch   0 Batch 1084/2154 - Train Accuracy: 0.5852, Validation Accuracy: 0.5902, Loss: 0.6390
Epoch   0 Batch 1088/2154 - Train Accuracy: 0.5750, Validation Accuracy: 0.5938, Loss: 0.6283
Epoch   0 Batch 1092/2154 - Train Accuracy: 0.6164, Validation Accuracy: 0.6058, Loss: 0.6592
Epoch   0 Batch 1096/2154 - Train Accuracy: 0.5773, Validation Accuracy: 0.6080, Loss: 0.6618
Epoch   0 Batch 1100/2154 - Train Accuracy: 0.5273, Validati

Epoch   0 Batch 1412/2154 - Train Accuracy: 0.6480, Validation Accuracy: 0.6165, Loss: 0.5475
Epoch   0 Batch 1416/2154 - Train Accuracy: 0.6241, Validation Accuracy: 0.6143, Loss: 0.5610
Epoch   0 Batch 1420/2154 - Train Accuracy: 0.6382, Validation Accuracy: 0.6037, Loss: 0.5552
Epoch   0 Batch 1424/2154 - Train Accuracy: 0.6570, Validation Accuracy: 0.5973, Loss: 0.5310
Epoch   0 Batch 1428/2154 - Train Accuracy: 0.6275, Validation Accuracy: 0.6165, Loss: 0.5630
Epoch   0 Batch 1432/2154 - Train Accuracy: 0.6406, Validation Accuracy: 0.5866, Loss: 0.5115
Epoch   0 Batch 1436/2154 - Train Accuracy: 0.6176, Validation Accuracy: 0.6193, Loss: 0.5263
Epoch   0 Batch 1440/2154 - Train Accuracy: 0.6094, Validation Accuracy: 0.6023, Loss: 0.5488
Epoch   0 Batch 1444/2154 - Train Accuracy: 0.6406, Validation Accuracy: 0.6087, Loss: 0.5323
Epoch   0 Batch 1448/2154 - Train Accuracy: 0.6503, Validation Accuracy: 0.6122, Loss: 0.5022
Epoch   0 Batch 1452/2154 - Train Accuracy: 0.6258, Validati

Epoch   0 Batch 1764/2154 - Train Accuracy: 0.6375, Validation Accuracy: 0.6406, Loss: 0.4452
Epoch   0 Batch 1768/2154 - Train Accuracy: 0.6176, Validation Accuracy: 0.6449, Loss: 0.5288
Epoch   0 Batch 1772/2154 - Train Accuracy: 0.6555, Validation Accuracy: 0.6484, Loss: 0.4098
Epoch   0 Batch 1776/2154 - Train Accuracy: 0.6595, Validation Accuracy: 0.6484, Loss: 0.4904
Epoch   0 Batch 1780/2154 - Train Accuracy: 0.6906, Validation Accuracy: 0.6364, Loss: 0.4101
Epoch   0 Batch 1784/2154 - Train Accuracy: 0.6513, Validation Accuracy: 0.6214, Loss: 0.4117
Epoch   0 Batch 1788/2154 - Train Accuracy: 0.7039, Validation Accuracy: 0.6470, Loss: 0.4204
Epoch   0 Batch 1792/2154 - Train Accuracy: 0.6953, Validation Accuracy: 0.6314, Loss: 0.4165
Epoch   0 Batch 1796/2154 - Train Accuracy: 0.6815, Validation Accuracy: 0.6278, Loss: 0.3999
Epoch   0 Batch 1800/2154 - Train Accuracy: 0.6797, Validation Accuracy: 0.6172, Loss: 0.4503
Epoch   0 Batch 1804/2154 - Train Accuracy: 0.7016, Validati

Epoch   0 Batch 2116/2154 - Train Accuracy: 0.6472, Validation Accuracy: 0.7138, Loss: 0.3796
Epoch   0 Batch 2120/2154 - Train Accuracy: 0.7122, Validation Accuracy: 0.6854, Loss: 0.3725
Epoch   0 Batch 2124/2154 - Train Accuracy: 0.6768, Validation Accuracy: 0.6811, Loss: 0.3698
Epoch   0 Batch 2128/2154 - Train Accuracy: 0.6933, Validation Accuracy: 0.6967, Loss: 0.3778
Epoch   0 Batch 2132/2154 - Train Accuracy: 0.6672, Validation Accuracy: 0.6854, Loss: 0.3686
Epoch   0 Batch 2136/2154 - Train Accuracy: 0.7570, Validation Accuracy: 0.6825, Loss: 0.3567
Epoch   0 Batch 2140/2154 - Train Accuracy: 0.7582, Validation Accuracy: 0.7124, Loss: 0.3071
Epoch   0 Batch 2144/2154 - Train Accuracy: 0.6752, Validation Accuracy: 0.7053, Loss: 0.3848
Epoch   0 Batch 2148/2154 - Train Accuracy: 0.7070, Validation Accuracy: 0.6960, Loss: 0.3904
Epoch   0 Batch 2152/2154 - Train Accuracy: 0.7118, Validation Accuracy: 0.6967, Loss: 0.3973
Epoch   1 Batch    4/2154 - Train Accuracy: 0.6883, Validati

Epoch   1 Batch  316/2154 - Train Accuracy: 0.7867, Validation Accuracy: 0.7095, Loss: 0.3086
Epoch   1 Batch  320/2154 - Train Accuracy: 0.7414, Validation Accuracy: 0.7195, Loss: 0.2863
Epoch   1 Batch  324/2154 - Train Accuracy: 0.7164, Validation Accuracy: 0.7060, Loss: 0.2966
Epoch   1 Batch  328/2154 - Train Accuracy: 0.7212, Validation Accuracy: 0.6982, Loss: 0.3289
Epoch   1 Batch  332/2154 - Train Accuracy: 0.7109, Validation Accuracy: 0.7003, Loss: 0.3109
Epoch   1 Batch  336/2154 - Train Accuracy: 0.7227, Validation Accuracy: 0.7138, Loss: 0.3531
Epoch   1 Batch  340/2154 - Train Accuracy: 0.7180, Validation Accuracy: 0.7109, Loss: 0.2914
Epoch   1 Batch  344/2154 - Train Accuracy: 0.6984, Validation Accuracy: 0.7145, Loss: 0.3083
Epoch   1 Batch  348/2154 - Train Accuracy: 0.6776, Validation Accuracy: 0.7131, Loss: 0.3253
Epoch   1 Batch  352/2154 - Train Accuracy: 0.7673, Validation Accuracy: 0.7010, Loss: 0.2977
Epoch   1 Batch  356/2154 - Train Accuracy: 0.7475, Validati

Epoch   1 Batch  668/2154 - Train Accuracy: 0.7977, Validation Accuracy: 0.7450, Loss: 0.2896
Epoch   1 Batch  672/2154 - Train Accuracy: 0.7932, Validation Accuracy: 0.7422, Loss: 0.2134
Epoch   1 Batch  676/2154 - Train Accuracy: 0.7541, Validation Accuracy: 0.7152, Loss: 0.2987
Epoch   1 Batch  680/2154 - Train Accuracy: 0.7253, Validation Accuracy: 0.7003, Loss: 0.2636
Epoch   1 Batch  684/2154 - Train Accuracy: 0.7714, Validation Accuracy: 0.7230, Loss: 0.3297
Epoch   1 Batch  688/2154 - Train Accuracy: 0.7195, Validation Accuracy: 0.7173, Loss: 0.2562
Epoch   1 Batch  692/2154 - Train Accuracy: 0.8237, Validation Accuracy: 0.7365, Loss: 0.2400
Epoch   1 Batch  696/2154 - Train Accuracy: 0.8289, Validation Accuracy: 0.7557, Loss: 0.2249
Epoch   1 Batch  700/2154 - Train Accuracy: 0.7599, Validation Accuracy: 0.7351, Loss: 0.2464
Epoch   1 Batch  704/2154 - Train Accuracy: 0.7286, Validation Accuracy: 0.7457, Loss: 0.2881
Epoch   1 Batch  708/2154 - Train Accuracy: 0.6957, Validati

Epoch   1 Batch 1020/2154 - Train Accuracy: 0.8133, Validation Accuracy: 0.7322, Loss: 0.2455
Epoch   1 Batch 1024/2154 - Train Accuracy: 0.7722, Validation Accuracy: 0.7592, Loss: 0.2186
Epoch   1 Batch 1028/2154 - Train Accuracy: 0.8174, Validation Accuracy: 0.7379, Loss: 0.2436
Epoch   1 Batch 1032/2154 - Train Accuracy: 0.7984, Validation Accuracy: 0.7457, Loss: 0.2045
Epoch   1 Batch 1036/2154 - Train Accuracy: 0.7747, Validation Accuracy: 0.7393, Loss: 0.2548
Epoch   1 Batch 1040/2154 - Train Accuracy: 0.7992, Validation Accuracy: 0.7528, Loss: 0.2011
Epoch   1 Batch 1044/2154 - Train Accuracy: 0.7609, Validation Accuracy: 0.7372, Loss: 0.2315
Epoch   1 Batch 1048/2154 - Train Accuracy: 0.8055, Validation Accuracy: 0.7536, Loss: 0.2275
Epoch   1 Batch 1052/2154 - Train Accuracy: 0.8503, Validation Accuracy: 0.7386, Loss: 0.2240
Epoch   1 Batch 1056/2154 - Train Accuracy: 0.7632, Validation Accuracy: 0.7337, Loss: 0.2130
Epoch   1 Batch 1060/2154 - Train Accuracy: 0.7617, Validati

Epoch   1 Batch 1372/2154 - Train Accuracy: 0.7852, Validation Accuracy: 0.7663, Loss: 0.1938
Epoch   1 Batch 1376/2154 - Train Accuracy: 0.8422, Validation Accuracy: 0.7351, Loss: 0.2088
Epoch   1 Batch 1380/2154 - Train Accuracy: 0.7781, Validation Accuracy: 0.7507, Loss: 0.1947
Epoch   1 Batch 1384/2154 - Train Accuracy: 0.7697, Validation Accuracy: 0.7415, Loss: 0.2366
Epoch   1 Batch 1388/2154 - Train Accuracy: 0.7418, Validation Accuracy: 0.7408, Loss: 0.2506
Epoch   1 Batch 1392/2154 - Train Accuracy: 0.8047, Validation Accuracy: 0.7663, Loss: 0.1785
Epoch   1 Batch 1396/2154 - Train Accuracy: 0.8469, Validation Accuracy: 0.7578, Loss: 0.1713
Epoch   1 Batch 1400/2154 - Train Accuracy: 0.8421, Validation Accuracy: 0.7642, Loss: 0.1922
Epoch   1 Batch 1404/2154 - Train Accuracy: 0.7582, Validation Accuracy: 0.7543, Loss: 0.2449
Epoch   1 Batch 1408/2154 - Train Accuracy: 0.8086, Validation Accuracy: 0.7884, Loss: 0.2164
Epoch   1 Batch 1412/2154 - Train Accuracy: 0.8298, Validati

Epoch   1 Batch 1724/2154 - Train Accuracy: 0.7969, Validation Accuracy: 0.7898, Loss: 0.1654
Epoch   1 Batch 1728/2154 - Train Accuracy: 0.8766, Validation Accuracy: 0.7969, Loss: 0.1521
Epoch   1 Batch 1732/2154 - Train Accuracy: 0.7983, Validation Accuracy: 0.7869, Loss: 0.1514
Epoch   1 Batch 1736/2154 - Train Accuracy: 0.8219, Validation Accuracy: 0.7940, Loss: 0.2167
Epoch   1 Batch 1740/2154 - Train Accuracy: 0.8528, Validation Accuracy: 0.7678, Loss: 0.1798
Epoch   1 Batch 1744/2154 - Train Accuracy: 0.8422, Validation Accuracy: 0.8018, Loss: 0.1721
Epoch   1 Batch 1748/2154 - Train Accuracy: 0.7266, Validation Accuracy: 0.7884, Loss: 0.2012
Epoch   1 Batch 1752/2154 - Train Accuracy: 0.7689, Validation Accuracy: 0.8054, Loss: 0.2232
Epoch   1 Batch 1756/2154 - Train Accuracy: 0.8258, Validation Accuracy: 0.7820, Loss: 0.1513
Epoch   1 Batch 1760/2154 - Train Accuracy: 0.8438, Validation Accuracy: 0.8196, Loss: 0.1594
Epoch   1 Batch 1764/2154 - Train Accuracy: 0.8172, Validati

Epoch   1 Batch 2076/2154 - Train Accuracy: 0.7789, Validation Accuracy: 0.8281, Loss: 0.1845
Epoch   1 Batch 2080/2154 - Train Accuracy: 0.8512, Validation Accuracy: 0.8139, Loss: 0.1481
Epoch   1 Batch 2084/2154 - Train Accuracy: 0.8867, Validation Accuracy: 0.8324, Loss: 0.1785
Epoch   1 Batch 2088/2154 - Train Accuracy: 0.8617, Validation Accuracy: 0.8175, Loss: 0.1783
Epoch   1 Batch 2092/2154 - Train Accuracy: 0.8313, Validation Accuracy: 0.8168, Loss: 0.1577
Epoch   1 Batch 2096/2154 - Train Accuracy: 0.8289, Validation Accuracy: 0.8082, Loss: 0.1480
Epoch   1 Batch 2100/2154 - Train Accuracy: 0.8719, Validation Accuracy: 0.8189, Loss: 0.1425
Epoch   1 Batch 2104/2154 - Train Accuracy: 0.8743, Validation Accuracy: 0.8345, Loss: 0.1560
Epoch   1 Batch 2108/2154 - Train Accuracy: 0.8348, Validation Accuracy: 0.8501, Loss: 0.1287
Epoch   1 Batch 2112/2154 - Train Accuracy: 0.8635, Validation Accuracy: 0.8501, Loss: 0.1560
Epoch   1 Batch 2116/2154 - Train Accuracy: 0.8676, Validati

Epoch   2 Batch  276/2154 - Train Accuracy: 0.8385, Validation Accuracy: 0.8082, Loss: 0.1278
Epoch   2 Batch  280/2154 - Train Accuracy: 0.8232, Validation Accuracy: 0.8374, Loss: 0.1797
Epoch   2 Batch  284/2154 - Train Accuracy: 0.8133, Validation Accuracy: 0.8139, Loss: 0.1667
Epoch   2 Batch  288/2154 - Train Accuracy: 0.8352, Validation Accuracy: 0.8246, Loss: 0.1546
Epoch   2 Batch  292/2154 - Train Accuracy: 0.8164, Validation Accuracy: 0.8530, Loss: 0.1564
Epoch   2 Batch  296/2154 - Train Accuracy: 0.8183, Validation Accuracy: 0.8118, Loss: 0.1859
Epoch   2 Batch  300/2154 - Train Accuracy: 0.8035, Validation Accuracy: 0.8189, Loss: 0.1590
Epoch   2 Batch  304/2154 - Train Accuracy: 0.8170, Validation Accuracy: 0.8111, Loss: 0.1481
Epoch   2 Batch  308/2154 - Train Accuracy: 0.8328, Validation Accuracy: 0.8281, Loss: 0.1476
Epoch   2 Batch  312/2154 - Train Accuracy: 0.8388, Validation Accuracy: 0.8189, Loss: 0.1528
Epoch   2 Batch  316/2154 - Train Accuracy: 0.8422, Validati

Epoch   2 Batch  628/2154 - Train Accuracy: 0.9046, Validation Accuracy: 0.8771, Loss: 0.1172
Epoch   2 Batch  632/2154 - Train Accuracy: 0.8333, Validation Accuracy: 0.8757, Loss: 0.1293
Epoch   2 Batch  636/2154 - Train Accuracy: 0.9071, Validation Accuracy: 0.8665, Loss: 0.1617
Epoch   2 Batch  640/2154 - Train Accuracy: 0.8500, Validation Accuracy: 0.8580, Loss: 0.1328
Epoch   2 Batch  644/2154 - Train Accuracy: 0.8187, Validation Accuracy: 0.8423, Loss: 0.1513
Epoch   2 Batch  648/2154 - Train Accuracy: 0.8656, Validation Accuracy: 0.8352, Loss: 0.1211
Epoch   2 Batch  652/2154 - Train Accuracy: 0.8380, Validation Accuracy: 0.8295, Loss: 0.1252
Epoch   2 Batch  656/2154 - Train Accuracy: 0.8195, Validation Accuracy: 0.8658, Loss: 0.1543
Epoch   2 Batch  660/2154 - Train Accuracy: 0.8469, Validation Accuracy: 0.8366, Loss: 0.1277
Epoch   2 Batch  664/2154 - Train Accuracy: 0.8717, Validation Accuracy: 0.8168, Loss: 0.1353
Epoch   2 Batch  668/2154 - Train Accuracy: 0.8709, Validati

Epoch   2 Batch  980/2154 - Train Accuracy: 0.8680, Validation Accuracy: 0.8658, Loss: 0.1098
Epoch   2 Batch  984/2154 - Train Accuracy: 0.8273, Validation Accuracy: 0.8537, Loss: 0.1524
Epoch   2 Batch  988/2154 - Train Accuracy: 0.8996, Validation Accuracy: 0.8679, Loss: 0.1058
Epoch   2 Batch  992/2154 - Train Accuracy: 0.8273, Validation Accuracy: 0.8729, Loss: 0.1249
Epoch   2 Batch  996/2154 - Train Accuracy: 0.8298, Validation Accuracy: 0.8565, Loss: 0.1403
Epoch   2 Batch 1000/2154 - Train Accuracy: 0.8787, Validation Accuracy: 0.8729, Loss: 0.1090
Epoch   2 Batch 1004/2154 - Train Accuracy: 0.8914, Validation Accuracy: 0.9006, Loss: 0.1149
Epoch   2 Batch 1008/2154 - Train Accuracy: 0.8938, Validation Accuracy: 0.8714, Loss: 0.1122
Epoch   2 Batch 1012/2154 - Train Accuracy: 0.8445, Validation Accuracy: 0.8771, Loss: 0.1087
Epoch   2 Batch 1016/2154 - Train Accuracy: 0.8477, Validation Accuracy: 0.8793, Loss: 0.1212
Epoch   2 Batch 1020/2154 - Train Accuracy: 0.8797, Validati

Epoch   2 Batch 1332/2154 - Train Accuracy: 0.8555, Validation Accuracy: 0.8892, Loss: 0.1033
Epoch   2 Batch 1336/2154 - Train Accuracy: 0.8495, Validation Accuracy: 0.9006, Loss: 0.1108
Epoch   2 Batch 1340/2154 - Train Accuracy: 0.8719, Validation Accuracy: 0.8963, Loss: 0.1306
Epoch   2 Batch 1344/2154 - Train Accuracy: 0.9186, Validation Accuracy: 0.8750, Loss: 0.1082
Epoch   2 Batch 1348/2154 - Train Accuracy: 0.8766, Validation Accuracy: 0.8615, Loss: 0.1040
Epoch   2 Batch 1352/2154 - Train Accuracy: 0.8984, Validation Accuracy: 0.8920, Loss: 0.1235
Epoch   2 Batch 1356/2154 - Train Accuracy: 0.8469, Validation Accuracy: 0.8963, Loss: 0.1400
Epoch   2 Batch 1360/2154 - Train Accuracy: 0.8281, Validation Accuracy: 0.8864, Loss: 0.0991
Epoch   2 Batch 1364/2154 - Train Accuracy: 0.9133, Validation Accuracy: 0.8892, Loss: 0.1289
Epoch   2 Batch 1368/2154 - Train Accuracy: 0.8906, Validation Accuracy: 0.8764, Loss: 0.1073
Epoch   2 Batch 1372/2154 - Train Accuracy: 0.8078, Validati

Epoch   2 Batch 1684/2154 - Train Accuracy: 0.9000, Validation Accuracy: 0.8920, Loss: 0.1195
Epoch   2 Batch 1688/2154 - Train Accuracy: 0.8943, Validation Accuracy: 0.9013, Loss: 0.0899
Epoch   2 Batch 1692/2154 - Train Accuracy: 0.8594, Validation Accuracy: 0.9070, Loss: 0.1112
Epoch   2 Batch 1696/2154 - Train Accuracy: 0.8773, Validation Accuracy: 0.8857, Loss: 0.1340
Epoch   2 Batch 1700/2154 - Train Accuracy: 0.8414, Validation Accuracy: 0.8771, Loss: 0.1106
Epoch   2 Batch 1704/2154 - Train Accuracy: 0.8690, Validation Accuracy: 0.8849, Loss: 0.1294
Epoch   2 Batch 1708/2154 - Train Accuracy: 0.8711, Validation Accuracy: 0.8707, Loss: 0.0978
Epoch   2 Batch 1712/2154 - Train Accuracy: 0.8462, Validation Accuracy: 0.8700, Loss: 0.0927
Epoch   2 Batch 1716/2154 - Train Accuracy: 0.8727, Validation Accuracy: 0.8686, Loss: 0.1002
Epoch   2 Batch 1720/2154 - Train Accuracy: 0.9104, Validation Accuracy: 0.8707, Loss: 0.1215
Epoch   2 Batch 1724/2154 - Train Accuracy: 0.8578, Validati

Epoch   2 Batch 2036/2154 - Train Accuracy: 0.8512, Validation Accuracy: 0.8707, Loss: 0.1104
Epoch   2 Batch 2040/2154 - Train Accuracy: 0.8816, Validation Accuracy: 0.8608, Loss: 0.1196
Epoch   2 Batch 2044/2154 - Train Accuracy: 0.8832, Validation Accuracy: 0.8487, Loss: 0.0896
Epoch   2 Batch 2048/2154 - Train Accuracy: 0.8714, Validation Accuracy: 0.8764, Loss: 0.0998
Epoch   2 Batch 2052/2154 - Train Accuracy: 0.9062, Validation Accuracy: 0.8977, Loss: 0.1049
Epoch   2 Batch 2056/2154 - Train Accuracy: 0.9336, Validation Accuracy: 0.9020, Loss: 0.0853
Epoch   2 Batch 2060/2154 - Train Accuracy: 0.8865, Validation Accuracy: 0.9112, Loss: 0.0947
Epoch   2 Batch 2064/2154 - Train Accuracy: 0.8964, Validation Accuracy: 0.9105, Loss: 0.1073
Epoch   2 Batch 2068/2154 - Train Accuracy: 0.8638, Validation Accuracy: 0.9027, Loss: 0.1048
Epoch   2 Batch 2072/2154 - Train Accuracy: 0.9062, Validation Accuracy: 0.8800, Loss: 0.0749
Epoch   2 Batch 2076/2154 - Train Accuracy: 0.8734, Validati

Epoch   3 Batch  236/2154 - Train Accuracy: 0.8684, Validation Accuracy: 0.9176, Loss: 0.0873
Epoch   3 Batch  240/2154 - Train Accuracy: 0.8711, Validation Accuracy: 0.8551, Loss: 0.1121
Epoch   3 Batch  244/2154 - Train Accuracy: 0.8844, Validation Accuracy: 0.8928, Loss: 0.0992
Epoch   3 Batch  248/2154 - Train Accuracy: 0.9000, Validation Accuracy: 0.9006, Loss: 0.0694
Epoch   3 Batch  252/2154 - Train Accuracy: 0.8867, Validation Accuracy: 0.8693, Loss: 0.1267
Epoch   3 Batch  256/2154 - Train Accuracy: 0.8492, Validation Accuracy: 0.8601, Loss: 0.0849
Epoch   3 Batch  260/2154 - Train Accuracy: 0.9187, Validation Accuracy: 0.8615, Loss: 0.0980
Epoch   3 Batch  264/2154 - Train Accuracy: 0.8586, Validation Accuracy: 0.8757, Loss: 0.0877
Epoch   3 Batch  268/2154 - Train Accuracy: 0.9169, Validation Accuracy: 0.9020, Loss: 0.0711
Epoch   3 Batch  272/2154 - Train Accuracy: 0.9071, Validation Accuracy: 0.9070, Loss: 0.0857
Epoch   3 Batch  276/2154 - Train Accuracy: 0.8624, Validati

Epoch   3 Batch  588/2154 - Train Accuracy: 0.8562, Validation Accuracy: 0.8977, Loss: 0.1007
Epoch   3 Batch  592/2154 - Train Accuracy: 0.8668, Validation Accuracy: 0.9162, Loss: 0.1034
Epoch   3 Batch  596/2154 - Train Accuracy: 0.8470, Validation Accuracy: 0.8835, Loss: 0.1217
Epoch   3 Batch  600/2154 - Train Accuracy: 0.9227, Validation Accuracy: 0.9034, Loss: 0.0684
Epoch   3 Batch  604/2154 - Train Accuracy: 0.9285, Validation Accuracy: 0.8956, Loss: 0.0756
Epoch   3 Batch  608/2154 - Train Accuracy: 0.9234, Validation Accuracy: 0.8928, Loss: 0.0949
Epoch   3 Batch  612/2154 - Train Accuracy: 0.9242, Validation Accuracy: 0.8864, Loss: 0.0755
Epoch   3 Batch  616/2154 - Train Accuracy: 0.8961, Validation Accuracy: 0.8459, Loss: 0.0900
Epoch   3 Batch  620/2154 - Train Accuracy: 0.9048, Validation Accuracy: 0.8764, Loss: 0.0743
Epoch   3 Batch  624/2154 - Train Accuracy: 0.9010, Validation Accuracy: 0.9226, Loss: 0.0826
Epoch   3 Batch  628/2154 - Train Accuracy: 0.9342, Validati

Epoch   3 Batch  940/2154 - Train Accuracy: 0.9156, Validation Accuracy: 0.9212, Loss: 0.0820
Epoch   3 Batch  944/2154 - Train Accuracy: 0.9219, Validation Accuracy: 0.8991, Loss: 0.0729
Epoch   3 Batch  948/2154 - Train Accuracy: 0.9030, Validation Accuracy: 0.9134, Loss: 0.0871
Epoch   3 Batch  952/2154 - Train Accuracy: 0.8832, Validation Accuracy: 0.8928, Loss: 0.0879
Epoch   3 Batch  956/2154 - Train Accuracy: 0.9263, Validation Accuracy: 0.8885, Loss: 0.0775
Epoch   3 Batch  960/2154 - Train Accuracy: 0.8891, Validation Accuracy: 0.8707, Loss: 0.0961
Epoch   3 Batch  964/2154 - Train Accuracy: 0.8672, Validation Accuracy: 0.8750, Loss: 0.0806
Epoch   3 Batch  968/2154 - Train Accuracy: 0.8836, Validation Accuracy: 0.9062, Loss: 0.0960
Epoch   3 Batch  972/2154 - Train Accuracy: 0.9539, Validation Accuracy: 0.8928, Loss: 0.0875
Epoch   3 Batch  976/2154 - Train Accuracy: 0.9285, Validation Accuracy: 0.8842, Loss: 0.0774
Epoch   3 Batch  980/2154 - Train Accuracy: 0.8891, Validati

Epoch   3 Batch 1292/2154 - Train Accuracy: 0.8923, Validation Accuracy: 0.9311, Loss: 0.0890
Epoch   3 Batch 1296/2154 - Train Accuracy: 0.8969, Validation Accuracy: 0.9134, Loss: 0.0742
Epoch   3 Batch 1300/2154 - Train Accuracy: 0.8727, Validation Accuracy: 0.8977, Loss: 0.0807
Epoch   3 Batch 1304/2154 - Train Accuracy: 0.9144, Validation Accuracy: 0.9219, Loss: 0.0724
Epoch   3 Batch 1308/2154 - Train Accuracy: 0.9219, Validation Accuracy: 0.9176, Loss: 0.0710
Epoch   3 Batch 1312/2154 - Train Accuracy: 0.9336, Validation Accuracy: 0.9176, Loss: 0.1016
Epoch   3 Batch 1316/2154 - Train Accuracy: 0.9523, Validation Accuracy: 0.8949, Loss: 0.0777
Epoch   3 Batch 1320/2154 - Train Accuracy: 0.9278, Validation Accuracy: 0.9055, Loss: 0.0573
Epoch   3 Batch 1324/2154 - Train Accuracy: 0.9133, Validation Accuracy: 0.9148, Loss: 0.0846
Epoch   3 Batch 1328/2154 - Train Accuracy: 0.8891, Validation Accuracy: 0.9098, Loss: 0.0721
Epoch   3 Batch 1332/2154 - Train Accuracy: 0.8750, Validati

Epoch   3 Batch 1644/2154 - Train Accuracy: 0.9219, Validation Accuracy: 0.9212, Loss: 0.0709
Epoch   3 Batch 1648/2154 - Train Accuracy: 0.8992, Validation Accuracy: 0.8984, Loss: 0.0728
Epoch   3 Batch 1652/2154 - Train Accuracy: 0.9320, Validation Accuracy: 0.8899, Loss: 0.0539
Epoch   3 Batch 1656/2154 - Train Accuracy: 0.9161, Validation Accuracy: 0.9091, Loss: 0.0619
Epoch   3 Batch 1660/2154 - Train Accuracy: 0.9350, Validation Accuracy: 0.9084, Loss: 0.0879
Epoch   3 Batch 1664/2154 - Train Accuracy: 0.8258, Validation Accuracy: 0.9212, Loss: 0.0844
Epoch   3 Batch 1668/2154 - Train Accuracy: 0.8977, Validation Accuracy: 0.9233, Loss: 0.0776
Epoch   3 Batch 1672/2154 - Train Accuracy: 0.9422, Validation Accuracy: 0.9233, Loss: 0.0492
Epoch   3 Batch 1676/2154 - Train Accuracy: 0.9273, Validation Accuracy: 0.9126, Loss: 0.0688
Epoch   3 Batch 1680/2154 - Train Accuracy: 0.9273, Validation Accuracy: 0.9148, Loss: 0.0588
Epoch   3 Batch 1684/2154 - Train Accuracy: 0.9164, Validati

Epoch   3 Batch 1996/2154 - Train Accuracy: 0.9219, Validation Accuracy: 0.8814, Loss: 0.0560
Epoch   3 Batch 2000/2154 - Train Accuracy: 0.9352, Validation Accuracy: 0.8622, Loss: 0.0726
Epoch   3 Batch 2004/2154 - Train Accuracy: 0.9226, Validation Accuracy: 0.8842, Loss: 0.0456
Epoch   3 Batch 2008/2154 - Train Accuracy: 0.9252, Validation Accuracy: 0.8970, Loss: 0.0577
Epoch   3 Batch 2012/2154 - Train Accuracy: 0.9227, Validation Accuracy: 0.8963, Loss: 0.0564
Epoch   3 Batch 2016/2154 - Train Accuracy: 0.9589, Validation Accuracy: 0.8928, Loss: 0.0756
Epoch   3 Batch 2020/2154 - Train Accuracy: 0.9539, Validation Accuracy: 0.8949, Loss: 0.0515
Epoch   3 Batch 2024/2154 - Train Accuracy: 0.9070, Validation Accuracy: 0.9070, Loss: 0.0564
Epoch   3 Batch 2028/2154 - Train Accuracy: 0.9141, Validation Accuracy: 0.9169, Loss: 0.0654
Epoch   3 Batch 2032/2154 - Train Accuracy: 0.8698, Validation Accuracy: 0.9091, Loss: 0.1039
Epoch   3 Batch 2036/2154 - Train Accuracy: 0.8832, Validati

Epoch   4 Batch  196/2154 - Train Accuracy: 0.9266, Validation Accuracy: 0.9020, Loss: 0.0592
Epoch   4 Batch  200/2154 - Train Accuracy: 0.8938, Validation Accuracy: 0.8956, Loss: 0.0604
Epoch   4 Batch  204/2154 - Train Accuracy: 0.9055, Validation Accuracy: 0.8899, Loss: 0.0581
Epoch   4 Batch  208/2154 - Train Accuracy: 0.9062, Validation Accuracy: 0.8828, Loss: 0.0659
Epoch   4 Batch  212/2154 - Train Accuracy: 0.9133, Validation Accuracy: 0.8807, Loss: 0.0657
Epoch   4 Batch  216/2154 - Train Accuracy: 0.8891, Validation Accuracy: 0.8878, Loss: 0.0725
Epoch   4 Batch  220/2154 - Train Accuracy: 0.9227, Validation Accuracy: 0.8949, Loss: 0.0631
Epoch   4 Batch  224/2154 - Train Accuracy: 0.9437, Validation Accuracy: 0.9084, Loss: 0.0644
Epoch   4 Batch  228/2154 - Train Accuracy: 0.8906, Validation Accuracy: 0.8956, Loss: 0.0655
Epoch   4 Batch  232/2154 - Train Accuracy: 0.8664, Validation Accuracy: 0.8892, Loss: 0.0606
Epoch   4 Batch  236/2154 - Train Accuracy: 0.8865, Validati

Epoch   4 Batch  548/2154 - Train Accuracy: 0.9342, Validation Accuracy: 0.8750, Loss: 0.0666
Epoch   4 Batch  552/2154 - Train Accuracy: 0.9129, Validation Accuracy: 0.8580, Loss: 0.0570
Epoch   4 Batch  556/2154 - Train Accuracy: 0.9546, Validation Accuracy: 0.8643, Loss: 0.0379
Epoch   4 Batch  560/2154 - Train Accuracy: 0.9148, Validation Accuracy: 0.8615, Loss: 0.0698
Epoch   4 Batch  564/2154 - Train Accuracy: 0.9336, Validation Accuracy: 0.8722, Loss: 0.0581
Epoch   4 Batch  568/2154 - Train Accuracy: 0.9281, Validation Accuracy: 0.8956, Loss: 0.0723
Epoch   4 Batch  572/2154 - Train Accuracy: 0.9211, Validation Accuracy: 0.8771, Loss: 0.0675
Epoch   4 Batch  576/2154 - Train Accuracy: 0.9555, Validation Accuracy: 0.9098, Loss: 0.0459
Epoch   4 Batch  580/2154 - Train Accuracy: 0.8988, Validation Accuracy: 0.9048, Loss: 0.0655
Epoch   4 Batch  584/2154 - Train Accuracy: 0.9144, Validation Accuracy: 0.9197, Loss: 0.0687
Epoch   4 Batch  588/2154 - Train Accuracy: 0.8844, Validati

Epoch   4 Batch  900/2154 - Train Accuracy: 0.9211, Validation Accuracy: 0.9020, Loss: 0.0868
Epoch   4 Batch  904/2154 - Train Accuracy: 0.9174, Validation Accuracy: 0.9162, Loss: 0.0732
Epoch   4 Batch  908/2154 - Train Accuracy: 0.9047, Validation Accuracy: 0.9148, Loss: 0.0587
Epoch   4 Batch  912/2154 - Train Accuracy: 0.9172, Validation Accuracy: 0.8771, Loss: 0.0718
Epoch   4 Batch  916/2154 - Train Accuracy: 0.8945, Validation Accuracy: 0.9162, Loss: 0.0697
Epoch   4 Batch  920/2154 - Train Accuracy: 0.9345, Validation Accuracy: 0.8963, Loss: 0.0435
Epoch   4 Batch  924/2154 - Train Accuracy: 0.9169, Validation Accuracy: 0.8793, Loss: 0.0485
Epoch   4 Batch  928/2154 - Train Accuracy: 0.9359, Validation Accuracy: 0.8942, Loss: 0.0620
Epoch   4 Batch  932/2154 - Train Accuracy: 0.9235, Validation Accuracy: 0.8871, Loss: 0.0821
Epoch   4 Batch  936/2154 - Train Accuracy: 0.9226, Validation Accuracy: 0.9162, Loss: 0.0682
Epoch   4 Batch  940/2154 - Train Accuracy: 0.9266, Validati

Epoch   4 Batch 1252/2154 - Train Accuracy: 0.9023, Validation Accuracy: 0.9027, Loss: 0.0700
Epoch   4 Batch 1256/2154 - Train Accuracy: 0.8969, Validation Accuracy: 0.9013, Loss: 0.0508
Epoch   4 Batch 1260/2154 - Train Accuracy: 0.8988, Validation Accuracy: 0.9169, Loss: 0.0969
Epoch   4 Batch 1264/2154 - Train Accuracy: 0.9117, Validation Accuracy: 0.9148, Loss: 0.0640
Epoch   4 Batch 1268/2154 - Train Accuracy: 0.9164, Validation Accuracy: 0.9027, Loss: 0.0631
Epoch   4 Batch 1272/2154 - Train Accuracy: 0.9095, Validation Accuracy: 0.9162, Loss: 0.0860
Epoch   4 Batch 1276/2154 - Train Accuracy: 0.9078, Validation Accuracy: 0.9226, Loss: 0.0555
Epoch   4 Batch 1280/2154 - Train Accuracy: 0.8849, Validation Accuracy: 0.9034, Loss: 0.0920
Epoch   4 Batch 1284/2154 - Train Accuracy: 0.8594, Validation Accuracy: 0.9453, Loss: 0.0695
Epoch   4 Batch 1288/2154 - Train Accuracy: 0.9204, Validation Accuracy: 0.9432, Loss: 0.0531
Epoch   4 Batch 1292/2154 - Train Accuracy: 0.9021, Validati

Epoch   4 Batch 1604/2154 - Train Accuracy: 0.9086, Validation Accuracy: 0.8352, Loss: 0.0602
Epoch   4 Batch 1608/2154 - Train Accuracy: 0.9515, Validation Accuracy: 0.8388, Loss: 0.0612
Epoch   4 Batch 1612/2154 - Train Accuracy: 0.9164, Validation Accuracy: 0.9027, Loss: 0.0617
Epoch   4 Batch 1616/2154 - Train Accuracy: 0.9062, Validation Accuracy: 0.9205, Loss: 0.0460
Epoch   4 Batch 1620/2154 - Train Accuracy: 0.9408, Validation Accuracy: 0.9254, Loss: 0.1119
Epoch   4 Batch 1624/2154 - Train Accuracy: 0.9334, Validation Accuracy: 0.9148, Loss: 0.0510
Epoch   4 Batch 1628/2154 - Train Accuracy: 0.9241, Validation Accuracy: 0.9041, Loss: 0.0569
Epoch   4 Batch 1632/2154 - Train Accuracy: 0.9211, Validation Accuracy: 0.8949, Loss: 0.0772
Epoch   4 Batch 1636/2154 - Train Accuracy: 0.9219, Validation Accuracy: 0.9006, Loss: 0.0593
Epoch   4 Batch 1640/2154 - Train Accuracy: 0.9617, Validation Accuracy: 0.9169, Loss: 0.0422
Epoch   4 Batch 1644/2154 - Train Accuracy: 0.9289, Validati

Epoch   4 Batch 1956/2154 - Train Accuracy: 0.9375, Validation Accuracy: 0.9119, Loss: 0.0385
Epoch   4 Batch 1960/2154 - Train Accuracy: 0.9169, Validation Accuracy: 0.9155, Loss: 0.0674
Epoch   4 Batch 1964/2154 - Train Accuracy: 0.9000, Validation Accuracy: 0.9212, Loss: 0.0608
Epoch   4 Batch 1968/2154 - Train Accuracy: 0.9079, Validation Accuracy: 0.9119, Loss: 0.0576
Epoch   4 Batch 1972/2154 - Train Accuracy: 0.9523, Validation Accuracy: 0.9212, Loss: 0.0506
Epoch   4 Batch 1976/2154 - Train Accuracy: 0.9196, Validation Accuracy: 0.9339, Loss: 0.0467
Epoch   4 Batch 1980/2154 - Train Accuracy: 0.9172, Validation Accuracy: 0.9396, Loss: 0.0714
Epoch   4 Batch 1984/2154 - Train Accuracy: 0.9156, Validation Accuracy: 0.9304, Loss: 0.0462
Epoch   4 Batch 1988/2154 - Train Accuracy: 0.9250, Validation Accuracy: 0.9126, Loss: 0.0532
Epoch   4 Batch 1992/2154 - Train Accuracy: 0.8996, Validation Accuracy: 0.9197, Loss: 0.0686
Epoch   4 Batch 1996/2154 - Train Accuracy: 0.9104, Validati

Epoch   5 Batch  156/2154 - Train Accuracy: 0.8914, Validation Accuracy: 0.8949, Loss: 0.0555
Epoch   5 Batch  160/2154 - Train Accuracy: 0.9250, Validation Accuracy: 0.8942, Loss: 0.0559
Epoch   5 Batch  164/2154 - Train Accuracy: 0.9202, Validation Accuracy: 0.8835, Loss: 0.0575
Epoch   5 Batch  168/2154 - Train Accuracy: 0.9400, Validation Accuracy: 0.9119, Loss: 0.0542
Epoch   5 Batch  172/2154 - Train Accuracy: 0.9071, Validation Accuracy: 0.9062, Loss: 0.0555
Epoch   5 Batch  176/2154 - Train Accuracy: 0.9523, Validation Accuracy: 0.9197, Loss: 0.0657
Epoch   5 Batch  180/2154 - Train Accuracy: 0.9117, Validation Accuracy: 0.8878, Loss: 0.0469
Epoch   5 Batch  184/2154 - Train Accuracy: 0.9572, Validation Accuracy: 0.8714, Loss: 0.0658
Epoch   5 Batch  188/2154 - Train Accuracy: 0.9305, Validation Accuracy: 0.8750, Loss: 0.0429
Epoch   5 Batch  192/2154 - Train Accuracy: 0.9256, Validation Accuracy: 0.8857, Loss: 0.0601
Epoch   5 Batch  196/2154 - Train Accuracy: 0.9047, Validati

Epoch   5 Batch  508/2154 - Train Accuracy: 0.9318, Validation Accuracy: 0.9325, Loss: 0.0615
Epoch   5 Batch  512/2154 - Train Accuracy: 0.8922, Validation Accuracy: 0.9318, Loss: 0.0553
Epoch   5 Batch  516/2154 - Train Accuracy: 0.9086, Validation Accuracy: 0.9183, Loss: 0.0612
Epoch   5 Batch  520/2154 - Train Accuracy: 0.8945, Validation Accuracy: 0.9077, Loss: 0.0537
Epoch   5 Batch  524/2154 - Train Accuracy: 0.9271, Validation Accuracy: 0.8892, Loss: 0.0487
Epoch   5 Batch  528/2154 - Train Accuracy: 0.9555, Validation Accuracy: 0.8935, Loss: 0.0482
Epoch   5 Batch  532/2154 - Train Accuracy: 0.9289, Validation Accuracy: 0.8928, Loss: 0.0460
Epoch   5 Batch  536/2154 - Train Accuracy: 0.8914, Validation Accuracy: 0.9126, Loss: 0.0529
Epoch   5 Batch  540/2154 - Train Accuracy: 0.9531, Validation Accuracy: 0.9219, Loss: 0.0464
Epoch   5 Batch  544/2154 - Train Accuracy: 0.9398, Validation Accuracy: 0.9155, Loss: 0.0422
Epoch   5 Batch  548/2154 - Train Accuracy: 0.9120, Validati

Epoch   5 Batch  860/2154 - Train Accuracy: 0.9070, Validation Accuracy: 0.9055, Loss: 0.0573
Epoch   5 Batch  864/2154 - Train Accuracy: 0.8859, Validation Accuracy: 0.8920, Loss: 0.0604
Epoch   5 Batch  868/2154 - Train Accuracy: 0.9406, Validation Accuracy: 0.9205, Loss: 0.0482
Epoch   5 Batch  872/2154 - Train Accuracy: 0.9490, Validation Accuracy: 0.8878, Loss: 0.0549
Epoch   5 Batch  876/2154 - Train Accuracy: 0.9523, Validation Accuracy: 0.9169, Loss: 0.0503
Epoch   5 Batch  880/2154 - Train Accuracy: 0.9016, Validation Accuracy: 0.9070, Loss: 0.0780
Epoch   5 Batch  884/2154 - Train Accuracy: 0.8852, Validation Accuracy: 0.9041, Loss: 0.0484
Epoch   5 Batch  888/2154 - Train Accuracy: 0.9655, Validation Accuracy: 0.9048, Loss: 0.0595
Epoch   5 Batch  892/2154 - Train Accuracy: 0.9449, Validation Accuracy: 0.9070, Loss: 0.0306
Epoch   5 Batch  896/2154 - Train Accuracy: 0.9414, Validation Accuracy: 0.9183, Loss: 0.0421
Epoch   5 Batch  900/2154 - Train Accuracy: 0.9367, Validati

Epoch   5 Batch 1212/2154 - Train Accuracy: 0.9178, Validation Accuracy: 0.9368, Loss: 0.0765
Epoch   5 Batch 1216/2154 - Train Accuracy: 0.9290, Validation Accuracy: 0.9084, Loss: 0.0530
Epoch   5 Batch 1220/2154 - Train Accuracy: 0.9477, Validation Accuracy: 0.9119, Loss: 0.0419
Epoch   5 Batch 1224/2154 - Train Accuracy: 0.9172, Validation Accuracy: 0.9027, Loss: 0.0576
Epoch   5 Batch 1228/2154 - Train Accuracy: 0.9148, Validation Accuracy: 0.8913, Loss: 0.0474
Epoch   5 Batch 1232/2154 - Train Accuracy: 0.9202, Validation Accuracy: 0.9020, Loss: 0.0393
Epoch   5 Batch 1236/2154 - Train Accuracy: 0.9227, Validation Accuracy: 0.9183, Loss: 0.0545
Epoch   5 Batch 1240/2154 - Train Accuracy: 0.8758, Validation Accuracy: 0.9205, Loss: 0.0560
Epoch   5 Batch 1244/2154 - Train Accuracy: 0.8875, Validation Accuracy: 0.8977, Loss: 0.0594
Epoch   5 Batch 1248/2154 - Train Accuracy: 0.9031, Validation Accuracy: 0.9062, Loss: 0.0740
Epoch   5 Batch 1252/2154 - Train Accuracy: 0.9172, Validati

Epoch   5 Batch 1564/2154 - Train Accuracy: 0.9554, Validation Accuracy: 0.9190, Loss: 0.0439
Epoch   5 Batch 1568/2154 - Train Accuracy: 0.9281, Validation Accuracy: 0.9084, Loss: 0.0531
Epoch   5 Batch 1572/2154 - Train Accuracy: 0.9196, Validation Accuracy: 0.9070, Loss: 0.0433
Epoch   5 Batch 1576/2154 - Train Accuracy: 0.9594, Validation Accuracy: 0.9375, Loss: 0.0636
Epoch   5 Batch 1580/2154 - Train Accuracy: 0.9161, Validation Accuracy: 0.9425, Loss: 0.0467
Epoch   5 Batch 1584/2154 - Train Accuracy: 0.9125, Validation Accuracy: 0.9332, Loss: 0.0389
Epoch   5 Batch 1588/2154 - Train Accuracy: 0.9625, Validation Accuracy: 0.9283, Loss: 0.0579
Epoch   5 Batch 1592/2154 - Train Accuracy: 0.9336, Validation Accuracy: 0.9084, Loss: 0.0616
Epoch   5 Batch 1596/2154 - Train Accuracy: 0.9164, Validation Accuracy: 0.9339, Loss: 0.0309
Epoch   5 Batch 1600/2154 - Train Accuracy: 0.8820, Validation Accuracy: 0.9318, Loss: 0.0593
Epoch   5 Batch 1604/2154 - Train Accuracy: 0.9273, Validati

Epoch   5 Batch 1916/2154 - Train Accuracy: 0.9397, Validation Accuracy: 0.9190, Loss: 0.0511
Epoch   5 Batch 1920/2154 - Train Accuracy: 0.9613, Validation Accuracy: 0.9325, Loss: 0.0395
Epoch   5 Batch 1924/2154 - Train Accuracy: 0.9008, Validation Accuracy: 0.9183, Loss: 0.0546
Epoch   5 Batch 1928/2154 - Train Accuracy: 0.9071, Validation Accuracy: 0.9254, Loss: 0.0724
Epoch   5 Batch 1932/2154 - Train Accuracy: 0.8882, Validation Accuracy: 0.9403, Loss: 0.0838
Epoch   5 Batch 1936/2154 - Train Accuracy: 0.9071, Validation Accuracy: 0.9219, Loss: 0.0548
Epoch   5 Batch 1940/2154 - Train Accuracy: 0.9033, Validation Accuracy: 0.9183, Loss: 0.0618
Epoch   5 Batch 1944/2154 - Train Accuracy: 0.9266, Validation Accuracy: 0.9432, Loss: 0.0654
Epoch   5 Batch 1948/2154 - Train Accuracy: 0.9383, Validation Accuracy: 0.9226, Loss: 0.0480
Epoch   5 Batch 1952/2154 - Train Accuracy: 0.9501, Validation Accuracy: 0.9212, Loss: 0.0398
Epoch   5 Batch 1956/2154 - Train Accuracy: 0.9367, Validati

Epoch   6 Batch  116/2154 - Train Accuracy: 0.8938, Validation Accuracy: 0.9162, Loss: 0.0641
Epoch   6 Batch  120/2154 - Train Accuracy: 0.9531, Validation Accuracy: 0.9141, Loss: 0.0504
Epoch   6 Batch  124/2154 - Train Accuracy: 0.9250, Validation Accuracy: 0.8942, Loss: 0.0602
Epoch   6 Batch  128/2154 - Train Accuracy: 0.9811, Validation Accuracy: 0.8729, Loss: 0.0330
Epoch   6 Batch  132/2154 - Train Accuracy: 0.9309, Validation Accuracy: 0.9283, Loss: 0.0397
Epoch   6 Batch  136/2154 - Train Accuracy: 0.9368, Validation Accuracy: 0.9354, Loss: 0.0401
Epoch   6 Batch  140/2154 - Train Accuracy: 0.9289, Validation Accuracy: 0.9411, Loss: 0.0611
Epoch   6 Batch  144/2154 - Train Accuracy: 0.9750, Validation Accuracy: 0.9254, Loss: 0.0218
Epoch   6 Batch  148/2154 - Train Accuracy: 0.9180, Validation Accuracy: 0.9148, Loss: 0.0476
Epoch   6 Batch  152/2154 - Train Accuracy: 0.9375, Validation Accuracy: 0.8906, Loss: 0.0485
Epoch   6 Batch  156/2154 - Train Accuracy: 0.9336, Validati

Epoch   6 Batch  468/2154 - Train Accuracy: 0.9211, Validation Accuracy: 0.9510, Loss: 0.0733
Epoch   6 Batch  472/2154 - Train Accuracy: 0.9092, Validation Accuracy: 0.9361, Loss: 0.0392
Epoch   6 Batch  476/2154 - Train Accuracy: 0.9382, Validation Accuracy: 0.8942, Loss: 0.0416
Epoch   6 Batch  480/2154 - Train Accuracy: 0.9546, Validation Accuracy: 0.8956, Loss: 0.0397
Epoch   6 Batch  484/2154 - Train Accuracy: 0.9679, Validation Accuracy: 0.9268, Loss: 0.0317
Epoch   6 Batch  488/2154 - Train Accuracy: 0.9117, Validation Accuracy: 0.9098, Loss: 0.0435
Epoch   6 Batch  492/2154 - Train Accuracy: 0.9524, Validation Accuracy: 0.9070, Loss: 0.0415
Epoch   6 Batch  496/2154 - Train Accuracy: 0.9227, Validation Accuracy: 0.9134, Loss: 0.0406
Epoch   6 Batch  500/2154 - Train Accuracy: 0.9770, Validation Accuracy: 0.9112, Loss: 0.0486
Epoch   6 Batch  504/2154 - Train Accuracy: 0.9018, Validation Accuracy: 0.9162, Loss: 0.0729
Epoch   6 Batch  508/2154 - Train Accuracy: 0.9112, Validati

Epoch   6 Batch  820/2154 - Train Accuracy: 0.9317, Validation Accuracy: 0.9446, Loss: 0.0590
Epoch   6 Batch  824/2154 - Train Accuracy: 0.9474, Validation Accuracy: 0.9403, Loss: 0.0635
Epoch   6 Batch  828/2154 - Train Accuracy: 0.9273, Validation Accuracy: 0.9460, Loss: 0.0446
Epoch   6 Batch  832/2154 - Train Accuracy: 0.9273, Validation Accuracy: 0.9482, Loss: 0.0328
Epoch   6 Batch  836/2154 - Train Accuracy: 0.9227, Validation Accuracy: 0.9268, Loss: 0.0622
Epoch   6 Batch  840/2154 - Train Accuracy: 0.9508, Validation Accuracy: 0.9212, Loss: 0.0523
Epoch   6 Batch  844/2154 - Train Accuracy: 0.9258, Validation Accuracy: 0.9162, Loss: 0.0750
Epoch   6 Batch  848/2154 - Train Accuracy: 0.9039, Validation Accuracy: 0.8935, Loss: 0.0608
Epoch   6 Batch  852/2154 - Train Accuracy: 0.9055, Validation Accuracy: 0.9041, Loss: 0.0347
Epoch   6 Batch  856/2154 - Train Accuracy: 0.8958, Validation Accuracy: 0.9176, Loss: 0.0443
Epoch   6 Batch  860/2154 - Train Accuracy: 0.9313, Validati

Epoch   6 Batch 1172/2154 - Train Accuracy: 0.9594, Validation Accuracy: 0.8928, Loss: 0.0319
Epoch   6 Batch 1176/2154 - Train Accuracy: 0.9435, Validation Accuracy: 0.9226, Loss: 0.0385
Epoch   6 Batch 1180/2154 - Train Accuracy: 0.8947, Validation Accuracy: 0.9055, Loss: 0.0507
Epoch   6 Batch 1184/2154 - Train Accuracy: 0.9119, Validation Accuracy: 0.9460, Loss: 0.0531
Epoch   6 Batch 1188/2154 - Train Accuracy: 0.8847, Validation Accuracy: 0.8935, Loss: 0.0630
Epoch   6 Batch 1192/2154 - Train Accuracy: 0.9449, Validation Accuracy: 0.9219, Loss: 0.0645
Epoch   6 Batch 1196/2154 - Train Accuracy: 0.9563, Validation Accuracy: 0.9318, Loss: 0.0428
Epoch   6 Batch 1200/2154 - Train Accuracy: 0.9186, Validation Accuracy: 0.9112, Loss: 0.0895
Epoch   6 Batch 1204/2154 - Train Accuracy: 0.9383, Validation Accuracy: 0.9247, Loss: 0.0492
Epoch   6 Batch 1208/2154 - Train Accuracy: 0.8996, Validation Accuracy: 0.9403, Loss: 0.0572
Epoch   6 Batch 1212/2154 - Train Accuracy: 0.9342, Validati

Epoch   6 Batch 1524/2154 - Train Accuracy: 0.8980, Validation Accuracy: 0.8942, Loss: 0.0449
Epoch   6 Batch 1528/2154 - Train Accuracy: 0.9196, Validation Accuracy: 0.8707, Loss: 0.0476
Epoch   6 Batch 1532/2154 - Train Accuracy: 0.9094, Validation Accuracy: 0.8913, Loss: 0.0566
Epoch   6 Batch 1536/2154 - Train Accuracy: 0.9252, Validation Accuracy: 0.9176, Loss: 0.0663
Epoch   6 Batch 1540/2154 - Train Accuracy: 0.9367, Validation Accuracy: 0.9197, Loss: 0.0480
Epoch   6 Batch 1544/2154 - Train Accuracy: 0.9734, Validation Accuracy: 0.9155, Loss: 0.0389
Epoch   6 Batch 1548/2154 - Train Accuracy: 0.9153, Validation Accuracy: 0.8942, Loss: 0.0364
Epoch   6 Batch 1552/2154 - Train Accuracy: 0.9602, Validation Accuracy: 0.9389, Loss: 0.0304
Epoch   6 Batch 1556/2154 - Train Accuracy: 0.9523, Validation Accuracy: 0.9268, Loss: 0.0428
Epoch   6 Batch 1560/2154 - Train Accuracy: 0.9016, Validation Accuracy: 0.9169, Loss: 0.0475
Epoch   6 Batch 1564/2154 - Train Accuracy: 0.9554, Validati

Epoch   6 Batch 1876/2154 - Train Accuracy: 0.9169, Validation Accuracy: 0.9212, Loss: 0.0485
Epoch   6 Batch 1880/2154 - Train Accuracy: 0.9000, Validation Accuracy: 0.9119, Loss: 0.0443
Epoch   6 Batch 1884/2154 - Train Accuracy: 0.9617, Validation Accuracy: 0.9098, Loss: 0.0341
Epoch   6 Batch 1888/2154 - Train Accuracy: 0.9868, Validation Accuracy: 0.9112, Loss: 0.0357
Epoch   6 Batch 1892/2154 - Train Accuracy: 0.9359, Validation Accuracy: 0.9155, Loss: 0.0352
Epoch   6 Batch 1896/2154 - Train Accuracy: 0.9120, Validation Accuracy: 0.9112, Loss: 0.0518
Epoch   6 Batch 1900/2154 - Train Accuracy: 0.9766, Validation Accuracy: 0.9041, Loss: 0.0442
Epoch   6 Batch 1904/2154 - Train Accuracy: 0.9435, Validation Accuracy: 0.9098, Loss: 0.0522
Epoch   6 Batch 1908/2154 - Train Accuracy: 0.9877, Validation Accuracy: 0.9212, Loss: 0.0247
Epoch   6 Batch 1912/2154 - Train Accuracy: 0.9883, Validation Accuracy: 0.9425, Loss: 0.0330
Epoch   6 Batch 1916/2154 - Train Accuracy: 0.9308, Validati

Epoch   7 Batch   76/2154 - Train Accuracy: 0.9625, Validation Accuracy: 0.9155, Loss: 0.0432
Epoch   7 Batch   80/2154 - Train Accuracy: 0.9109, Validation Accuracy: 0.8984, Loss: 0.0624
Epoch   7 Batch   84/2154 - Train Accuracy: 0.9241, Validation Accuracy: 0.9141, Loss: 0.0437
Epoch   7 Batch   88/2154 - Train Accuracy: 0.9704, Validation Accuracy: 0.9240, Loss: 0.0207
Epoch   7 Batch   92/2154 - Train Accuracy: 0.9195, Validation Accuracy: 0.9233, Loss: 0.0526
Epoch   7 Batch   96/2154 - Train Accuracy: 0.8980, Validation Accuracy: 0.9261, Loss: 0.0655
Epoch   7 Batch  100/2154 - Train Accuracy: 0.8936, Validation Accuracy: 0.9190, Loss: 0.0687
Epoch   7 Batch  104/2154 - Train Accuracy: 0.9474, Validation Accuracy: 0.9197, Loss: 0.0347
Epoch   7 Batch  108/2154 - Train Accuracy: 0.9350, Validation Accuracy: 0.9247, Loss: 0.0339
Epoch   7 Batch  112/2154 - Train Accuracy: 0.9070, Validation Accuracy: 0.8999, Loss: 0.0422
Epoch   7 Batch  116/2154 - Train Accuracy: 0.8984, Validati

Epoch   7 Batch  428/2154 - Train Accuracy: 0.9578, Validation Accuracy: 0.8835, Loss: 0.0260
Epoch   7 Batch  432/2154 - Train Accuracy: 0.9234, Validation Accuracy: 0.8821, Loss: 0.0482
Epoch   7 Batch  436/2154 - Train Accuracy: 0.9441, Validation Accuracy: 0.8935, Loss: 0.0366
Epoch   7 Batch  440/2154 - Train Accuracy: 0.9515, Validation Accuracy: 0.9411, Loss: 0.0364
Epoch   7 Batch  444/2154 - Train Accuracy: 0.9383, Validation Accuracy: 0.9489, Loss: 0.0321
Epoch   7 Batch  448/2154 - Train Accuracy: 0.8750, Validation Accuracy: 0.9446, Loss: 0.0496
Epoch   7 Batch  452/2154 - Train Accuracy: 0.9391, Validation Accuracy: 0.9361, Loss: 0.0499
Epoch   7 Batch  456/2154 - Train Accuracy: 0.9000, Validation Accuracy: 0.9098, Loss: 0.0459
Epoch   7 Batch  460/2154 - Train Accuracy: 0.8727, Validation Accuracy: 0.9474, Loss: 0.0505
Epoch   7 Batch  464/2154 - Train Accuracy: 0.9609, Validation Accuracy: 0.9283, Loss: 0.0687
Epoch   7 Batch  468/2154 - Train Accuracy: 0.9141, Validati

Epoch   7 Batch  780/2154 - Train Accuracy: 0.9406, Validation Accuracy: 0.9126, Loss: 0.0316
Epoch   7 Batch  784/2154 - Train Accuracy: 0.9194, Validation Accuracy: 0.9155, Loss: 0.0744
Epoch   7 Batch  788/2154 - Train Accuracy: 0.9301, Validation Accuracy: 0.9425, Loss: 0.0342
Epoch   7 Batch  792/2154 - Train Accuracy: 0.9243, Validation Accuracy: 0.9155, Loss: 0.0589
Epoch   7 Batch  796/2154 - Train Accuracy: 0.9021, Validation Accuracy: 0.9070, Loss: 0.0559
Epoch   7 Batch  800/2154 - Train Accuracy: 0.9548, Validation Accuracy: 0.9070, Loss: 0.0408
Epoch   7 Batch  804/2154 - Train Accuracy: 0.9609, Validation Accuracy: 0.9013, Loss: 0.0245
Epoch   7 Batch  808/2154 - Train Accuracy: 0.9441, Validation Accuracy: 0.9119, Loss: 0.0576
Epoch   7 Batch  812/2154 - Train Accuracy: 0.9465, Validation Accuracy: 0.9126, Loss: 0.0540
Epoch   7 Batch  816/2154 - Train Accuracy: 0.9326, Validation Accuracy: 0.9503, Loss: 0.0359
Epoch   7 Batch  820/2154 - Train Accuracy: 0.9531, Validati

Epoch   7 Batch 1132/2154 - Train Accuracy: 0.9360, Validation Accuracy: 0.9091, Loss: 0.0561
Epoch   7 Batch 1136/2154 - Train Accuracy: 0.9445, Validation Accuracy: 0.9190, Loss: 0.0331
Epoch   7 Batch 1140/2154 - Train Accuracy: 0.9586, Validation Accuracy: 0.9062, Loss: 0.0360
Epoch   7 Batch 1144/2154 - Train Accuracy: 0.9606, Validation Accuracy: 0.9254, Loss: 0.0302
Epoch   7 Batch 1148/2154 - Train Accuracy: 0.9490, Validation Accuracy: 0.9112, Loss: 0.0553
Epoch   7 Batch 1152/2154 - Train Accuracy: 0.9375, Validation Accuracy: 0.9283, Loss: 0.0312
Epoch   7 Batch 1156/2154 - Train Accuracy: 0.9350, Validation Accuracy: 0.9368, Loss: 0.0553
Epoch   7 Batch 1160/2154 - Train Accuracy: 0.9469, Validation Accuracy: 0.9119, Loss: 0.0338
Epoch   7 Batch 1164/2154 - Train Accuracy: 0.9141, Validation Accuracy: 0.9134, Loss: 0.0495
Epoch   7 Batch 1168/2154 - Train Accuracy: 0.9408, Validation Accuracy: 0.9197, Loss: 0.0441
Epoch   7 Batch 1172/2154 - Train Accuracy: 0.9633, Validati

Epoch   7 Batch 1484/2154 - Train Accuracy: 0.9539, Validation Accuracy: 0.9432, Loss: 0.0368
Epoch   7 Batch 1488/2154 - Train Accuracy: 0.9479, Validation Accuracy: 0.9425, Loss: 0.0507
Epoch   7 Batch 1492/2154 - Train Accuracy: 0.9398, Validation Accuracy: 0.9425, Loss: 0.0306
Epoch   7 Batch 1496/2154 - Train Accuracy: 0.9500, Validation Accuracy: 0.9219, Loss: 0.0324
Epoch   7 Batch 1500/2154 - Train Accuracy: 0.9589, Validation Accuracy: 0.9084, Loss: 0.0267
Epoch   7 Batch 1504/2154 - Train Accuracy: 0.9750, Validation Accuracy: 0.9304, Loss: 0.0315
Epoch   7 Batch 1508/2154 - Train Accuracy: 0.9437, Validation Accuracy: 0.9318, Loss: 0.0269
Epoch   7 Batch 1512/2154 - Train Accuracy: 0.9234, Validation Accuracy: 0.9197, Loss: 0.0546
Epoch   7 Batch 1516/2154 - Train Accuracy: 0.9317, Validation Accuracy: 0.9212, Loss: 0.0425
Epoch   7 Batch 1520/2154 - Train Accuracy: 0.9416, Validation Accuracy: 0.9347, Loss: 0.0314
Epoch   7 Batch 1524/2154 - Train Accuracy: 0.8972, Validati

Epoch   7 Batch 1836/2154 - Train Accuracy: 0.9711, Validation Accuracy: 0.9361, Loss: 0.0267
Epoch   7 Batch 1840/2154 - Train Accuracy: 0.9575, Validation Accuracy: 0.9489, Loss: 0.0262
Epoch   7 Batch 1844/2154 - Train Accuracy: 0.9109, Validation Accuracy: 0.9538, Loss: 0.0275
Epoch   7 Batch 1848/2154 - Train Accuracy: 0.9836, Validation Accuracy: 0.9311, Loss: 0.0210
Epoch   7 Batch 1852/2154 - Train Accuracy: 0.9315, Validation Accuracy: 0.9048, Loss: 0.0360
Epoch   7 Batch 1856/2154 - Train Accuracy: 0.8719, Validation Accuracy: 0.9183, Loss: 0.0549
Epoch   7 Batch 1860/2154 - Train Accuracy: 0.9350, Validation Accuracy: 0.9141, Loss: 0.0370
Epoch   7 Batch 1864/2154 - Train Accuracy: 0.9508, Validation Accuracy: 0.9460, Loss: 0.0220
Epoch   7 Batch 1868/2154 - Train Accuracy: 0.9186, Validation Accuracy: 0.9276, Loss: 0.0388
Epoch   7 Batch 1872/2154 - Train Accuracy: 0.9305, Validation Accuracy: 0.9354, Loss: 0.0440
Epoch   7 Batch 1876/2154 - Train Accuracy: 0.9153, Validati

Epoch   8 Batch   36/2154 - Train Accuracy: 0.9445, Validation Accuracy: 0.9538, Loss: 0.0431
Epoch   8 Batch   40/2154 - Train Accuracy: 0.9539, Validation Accuracy: 0.9325, Loss: 0.0391
Epoch   8 Batch   44/2154 - Train Accuracy: 0.9320, Validation Accuracy: 0.9368, Loss: 0.0534
Epoch   8 Batch   48/2154 - Train Accuracy: 0.9482, Validation Accuracy: 0.9361, Loss: 0.0410
Epoch   8 Batch   52/2154 - Train Accuracy: 0.9539, Validation Accuracy: 0.9418, Loss: 0.0256
Epoch   8 Batch   56/2154 - Train Accuracy: 0.9524, Validation Accuracy: 0.9126, Loss: 0.0420
Epoch   8 Batch   60/2154 - Train Accuracy: 0.9258, Validation Accuracy: 0.9418, Loss: 0.0414
Epoch   8 Batch   64/2154 - Train Accuracy: 0.9531, Validation Accuracy: 0.9545, Loss: 0.0393
Epoch   8 Batch   68/2154 - Train Accuracy: 0.9122, Validation Accuracy: 0.9545, Loss: 0.0386
Epoch   8 Batch   72/2154 - Train Accuracy: 0.9891, Validation Accuracy: 0.9339, Loss: 0.0247
Epoch   8 Batch   76/2154 - Train Accuracy: 0.9672, Validati

Epoch   8 Batch  388/2154 - Train Accuracy: 0.9219, Validation Accuracy: 0.9453, Loss: 0.0393
Epoch   8 Batch  392/2154 - Train Accuracy: 0.9492, Validation Accuracy: 0.9325, Loss: 0.0390
Epoch   8 Batch  396/2154 - Train Accuracy: 0.9235, Validation Accuracy: 0.9290, Loss: 0.0435
Epoch   8 Batch  400/2154 - Train Accuracy: 0.9625, Validation Accuracy: 0.9389, Loss: 0.0268
Epoch   8 Batch  404/2154 - Train Accuracy: 0.9570, Validation Accuracy: 0.9396, Loss: 0.0264
Epoch   8 Batch  408/2154 - Train Accuracy: 0.9336, Validation Accuracy: 0.9396, Loss: 0.0286
Epoch   8 Batch  412/2154 - Train Accuracy: 0.9391, Validation Accuracy: 0.9084, Loss: 0.0360
Epoch   8 Batch  416/2154 - Train Accuracy: 0.9617, Validation Accuracy: 0.9062, Loss: 0.0343
Epoch   8 Batch  420/2154 - Train Accuracy: 0.9487, Validation Accuracy: 0.9148, Loss: 0.0331
Epoch   8 Batch  424/2154 - Train Accuracy: 0.9186, Validation Accuracy: 0.8963, Loss: 0.0497
Epoch   8 Batch  428/2154 - Train Accuracy: 0.9594, Validati

Epoch   8 Batch  740/2154 - Train Accuracy: 0.9398, Validation Accuracy: 0.9467, Loss: 0.0445
Epoch   8 Batch  744/2154 - Train Accuracy: 0.9633, Validation Accuracy: 0.9432, Loss: 0.0367
Epoch   8 Batch  748/2154 - Train Accuracy: 0.9695, Validation Accuracy: 0.9339, Loss: 0.0182
Epoch   8 Batch  752/2154 - Train Accuracy: 0.9574, Validation Accuracy: 0.9347, Loss: 0.0441
Epoch   8 Batch  756/2154 - Train Accuracy: 0.9547, Validation Accuracy: 0.8977, Loss: 0.0288
Epoch   8 Batch  760/2154 - Train Accuracy: 0.9461, Validation Accuracy: 0.9027, Loss: 0.0363
Epoch   8 Batch  764/2154 - Train Accuracy: 0.9646, Validation Accuracy: 0.9268, Loss: 0.0343
Epoch   8 Batch  768/2154 - Train Accuracy: 0.9762, Validation Accuracy: 0.9098, Loss: 0.0305
Epoch   8 Batch  772/2154 - Train Accuracy: 0.9430, Validation Accuracy: 0.8963, Loss: 0.0231
Epoch   8 Batch  776/2154 - Train Accuracy: 0.9477, Validation Accuracy: 0.8906, Loss: 0.0291
Epoch   8 Batch  780/2154 - Train Accuracy: 0.9320, Validati

Epoch   8 Batch 1092/2154 - Train Accuracy: 0.9391, Validation Accuracy: 0.9510, Loss: 0.0445
Epoch   8 Batch 1096/2154 - Train Accuracy: 0.9589, Validation Accuracy: 0.9751, Loss: 0.0489
Epoch   8 Batch 1100/2154 - Train Accuracy: 0.9219, Validation Accuracy: 0.9631, Loss: 0.0401
Epoch   8 Batch 1104/2154 - Train Accuracy: 0.9445, Validation Accuracy: 0.9446, Loss: 0.0506
Epoch   8 Batch 1108/2154 - Train Accuracy: 0.9449, Validation Accuracy: 0.9517, Loss: 0.0582
Epoch   8 Batch 1112/2154 - Train Accuracy: 0.9578, Validation Accuracy: 0.9439, Loss: 0.0293
Epoch   8 Batch 1116/2154 - Train Accuracy: 0.9352, Validation Accuracy: 0.9482, Loss: 0.0398
Epoch   8 Batch 1120/2154 - Train Accuracy: 0.9641, Validation Accuracy: 0.9446, Loss: 0.0319
Epoch   8 Batch 1124/2154 - Train Accuracy: 0.9297, Validation Accuracy: 0.9702, Loss: 0.0370
Epoch   8 Batch 1128/2154 - Train Accuracy: 0.9094, Validation Accuracy: 0.9751, Loss: 0.0440
Epoch   8 Batch 1132/2154 - Train Accuracy: 0.9382, Validati

Epoch   8 Batch 1444/2154 - Train Accuracy: 0.9508, Validation Accuracy: 0.9425, Loss: 0.0335
Epoch   8 Batch 1448/2154 - Train Accuracy: 0.9315, Validation Accuracy: 0.9375, Loss: 0.0368
Epoch   8 Batch 1452/2154 - Train Accuracy: 0.9313, Validation Accuracy: 0.9254, Loss: 0.0348
Epoch   8 Batch 1456/2154 - Train Accuracy: 0.9786, Validation Accuracy: 0.9254, Loss: 0.0221
Epoch   8 Batch 1460/2154 - Train Accuracy: 0.9180, Validation Accuracy: 0.9197, Loss: 0.0537
Epoch   8 Batch 1464/2154 - Train Accuracy: 0.9219, Validation Accuracy: 0.9290, Loss: 0.0423
Epoch   8 Batch 1468/2154 - Train Accuracy: 0.9430, Validation Accuracy: 0.9347, Loss: 0.0291
Epoch   8 Batch 1472/2154 - Train Accuracy: 0.9430, Validation Accuracy: 0.9531, Loss: 0.0369
Epoch   8 Batch 1476/2154 - Train Accuracy: 0.9169, Validation Accuracy: 0.9489, Loss: 0.0346
Epoch   8 Batch 1480/2154 - Train Accuracy: 0.9359, Validation Accuracy: 0.9489, Loss: 0.0517
Epoch   8 Batch 1484/2154 - Train Accuracy: 0.9457, Validati

Epoch   8 Batch 1796/2154 - Train Accuracy: 0.9286, Validation Accuracy: 0.9411, Loss: 0.0429
Epoch   8 Batch 1800/2154 - Train Accuracy: 0.9570, Validation Accuracy: 0.9517, Loss: 0.0505
Epoch   8 Batch 1804/2154 - Train Accuracy: 0.9323, Validation Accuracy: 0.9517, Loss: 0.0336
Epoch   8 Batch 1808/2154 - Train Accuracy: 0.9161, Validation Accuracy: 0.9709, Loss: 0.0463
Epoch   8 Batch 1812/2154 - Train Accuracy: 0.9117, Validation Accuracy: 0.9737, Loss: 0.0335
Epoch   8 Batch 1816/2154 - Train Accuracy: 0.9470, Validation Accuracy: 0.9787, Loss: 0.0431
Epoch   8 Batch 1820/2154 - Train Accuracy: 0.9547, Validation Accuracy: 0.9787, Loss: 0.0288
Epoch   8 Batch 1824/2154 - Train Accuracy: 0.9162, Validation Accuracy: 0.9787, Loss: 0.0456
Epoch   8 Batch 1828/2154 - Train Accuracy: 0.9547, Validation Accuracy: 0.9787, Loss: 0.0600
Epoch   8 Batch 1832/2154 - Train Accuracy: 0.9663, Validation Accuracy: 0.9503, Loss: 0.0342
Epoch   8 Batch 1836/2154 - Train Accuracy: 0.9602, Validati

Epoch   8 Batch 2148/2154 - Train Accuracy: 0.9148, Validation Accuracy: 0.9368, Loss: 0.0448
Epoch   8 Batch 2152/2154 - Train Accuracy: 0.9462, Validation Accuracy: 0.9673, Loss: 0.0518
Epoch   9 Batch    4/2154 - Train Accuracy: 0.9875, Validation Accuracy: 0.9666, Loss: 0.0187
Epoch   9 Batch    8/2154 - Train Accuracy: 0.9457, Validation Accuracy: 0.9496, Loss: 0.0495
Epoch   9 Batch   12/2154 - Train Accuracy: 0.9367, Validation Accuracy: 0.9268, Loss: 0.0514
Epoch   9 Batch   16/2154 - Train Accuracy: 0.9602, Validation Accuracy: 0.9503, Loss: 0.0214
Epoch   9 Batch   20/2154 - Train Accuracy: 0.9406, Validation Accuracy: 0.9560, Loss: 0.0401
Epoch   9 Batch   24/2154 - Train Accuracy: 0.9062, Validation Accuracy: 0.9183, Loss: 0.0533
Epoch   9 Batch   28/2154 - Train Accuracy: 0.9286, Validation Accuracy: 0.9183, Loss: 0.0675
Epoch   9 Batch   32/2154 - Train Accuracy: 0.9309, Validation Accuracy: 0.9197, Loss: 0.0477
Epoch   9 Batch   36/2154 - Train Accuracy: 0.9117, Validati

Epoch   9 Batch  348/2154 - Train Accuracy: 0.9465, Validation Accuracy: 0.9105, Loss: 0.0359
Epoch   9 Batch  352/2154 - Train Accuracy: 0.9630, Validation Accuracy: 0.9297, Loss: 0.0260
Epoch   9 Batch  356/2154 - Train Accuracy: 0.9638, Validation Accuracy: 0.9517, Loss: 0.0259
Epoch   9 Batch  360/2154 - Train Accuracy: 0.9696, Validation Accuracy: 0.9616, Loss: 0.0328
Epoch   9 Batch  364/2154 - Train Accuracy: 0.9602, Validation Accuracy: 0.9368, Loss: 0.0302
Epoch   9 Batch  368/2154 - Train Accuracy: 0.9094, Validation Accuracy: 0.9276, Loss: 0.0372
Epoch   9 Batch  372/2154 - Train Accuracy: 0.9301, Validation Accuracy: 0.9190, Loss: 0.0533
Epoch   9 Batch  376/2154 - Train Accuracy: 0.9547, Validation Accuracy: 0.9389, Loss: 0.0258
Epoch   9 Batch  380/2154 - Train Accuracy: 0.9273, Validation Accuracy: 0.9361, Loss: 0.0336
Epoch   9 Batch  384/2154 - Train Accuracy: 0.8942, Validation Accuracy: 0.9567, Loss: 0.0342
Epoch   9 Batch  388/2154 - Train Accuracy: 0.9375, Validati

Epoch   9 Batch  700/2154 - Train Accuracy: 0.9572, Validation Accuracy: 0.9787, Loss: 0.0277
Epoch   9 Batch  704/2154 - Train Accuracy: 0.9441, Validation Accuracy: 0.9467, Loss: 0.0363
Epoch   9 Batch  708/2154 - Train Accuracy: 0.9457, Validation Accuracy: 0.9496, Loss: 0.0515
Epoch   9 Batch  712/2154 - Train Accuracy: 0.9375, Validation Accuracy: 0.9503, Loss: 0.0408
Epoch   9 Batch  716/2154 - Train Accuracy: 0.8929, Validation Accuracy: 0.9482, Loss: 0.0389
Epoch   9 Batch  720/2154 - Train Accuracy: 0.9630, Validation Accuracy: 0.9538, Loss: 0.0324
Epoch   9 Batch  724/2154 - Train Accuracy: 0.9613, Validation Accuracy: 0.9474, Loss: 0.0403
Epoch   9 Batch  728/2154 - Train Accuracy: 0.8688, Validation Accuracy: 0.9304, Loss: 0.0490
Epoch   9 Batch  732/2154 - Train Accuracy: 0.9770, Validation Accuracy: 0.9190, Loss: 0.0257
Epoch   9 Batch  736/2154 - Train Accuracy: 0.9492, Validation Accuracy: 0.9254, Loss: 0.0341
Epoch   9 Batch  740/2154 - Train Accuracy: 0.9555, Validati

Epoch   9 Batch 1052/2154 - Train Accuracy: 0.9605, Validation Accuracy: 0.9205, Loss: 0.0497
Epoch   9 Batch 1056/2154 - Train Accuracy: 0.9572, Validation Accuracy: 0.9496, Loss: 0.0300
Epoch   9 Batch 1060/2154 - Train Accuracy: 0.9164, Validation Accuracy: 0.9474, Loss: 0.0387
Epoch   9 Batch 1064/2154 - Train Accuracy: 0.9336, Validation Accuracy: 0.9247, Loss: 0.0413
Epoch   9 Batch 1068/2154 - Train Accuracy: 0.9474, Validation Accuracy: 0.9389, Loss: 0.0317
Epoch   9 Batch 1072/2154 - Train Accuracy: 0.9273, Validation Accuracy: 0.9467, Loss: 0.0398
Epoch   9 Batch 1076/2154 - Train Accuracy: 0.9243, Validation Accuracy: 0.9659, Loss: 0.0342
Epoch   9 Batch 1080/2154 - Train Accuracy: 0.9172, Validation Accuracy: 0.9730, Loss: 0.0525
Epoch   9 Batch 1084/2154 - Train Accuracy: 0.9258, Validation Accuracy: 0.9680, Loss: 0.0348
Epoch   9 Batch 1088/2154 - Train Accuracy: 0.9422, Validation Accuracy: 0.9609, Loss: 0.0248
Epoch   9 Batch 1092/2154 - Train Accuracy: 0.9461, Validati

Epoch   9 Batch 1404/2154 - Train Accuracy: 0.9515, Validation Accuracy: 0.9446, Loss: 0.0337
Epoch   9 Batch 1408/2154 - Train Accuracy: 0.9273, Validation Accuracy: 0.9396, Loss: 0.0484
Epoch   9 Batch 1412/2154 - Train Accuracy: 0.9424, Validation Accuracy: 0.9467, Loss: 0.0300
Epoch   9 Batch 1416/2154 - Train Accuracy: 0.9470, Validation Accuracy: 0.9531, Loss: 0.0337
Epoch   9 Batch 1420/2154 - Train Accuracy: 0.9309, Validation Accuracy: 0.9411, Loss: 0.0403
Epoch   9 Batch 1424/2154 - Train Accuracy: 0.8992, Validation Accuracy: 0.9254, Loss: 0.0623
Epoch   9 Batch 1428/2154 - Train Accuracy: 0.9219, Validation Accuracy: 0.9411, Loss: 0.0473
Epoch   9 Batch 1432/2154 - Train Accuracy: 0.9313, Validation Accuracy: 0.9489, Loss: 0.0278
Epoch   9 Batch 1436/2154 - Train Accuracy: 0.9597, Validation Accuracy: 0.9389, Loss: 0.0315
Epoch   9 Batch 1440/2154 - Train Accuracy: 0.9352, Validation Accuracy: 0.9588, Loss: 0.0316
Epoch   9 Batch 1444/2154 - Train Accuracy: 0.9289, Validati

Epoch   9 Batch 1756/2154 - Train Accuracy: 0.9609, Validation Accuracy: 0.9411, Loss: 0.0188
Epoch   9 Batch 1760/2154 - Train Accuracy: 0.9416, Validation Accuracy: 0.9425, Loss: 0.0323
Epoch   9 Batch 1764/2154 - Train Accuracy: 0.9586, Validation Accuracy: 0.9467, Loss: 0.0396
Epoch   9 Batch 1768/2154 - Train Accuracy: 0.8824, Validation Accuracy: 0.9403, Loss: 0.0534
Epoch   9 Batch 1772/2154 - Train Accuracy: 0.9883, Validation Accuracy: 0.9510, Loss: 0.0183
Epoch   9 Batch 1776/2154 - Train Accuracy: 0.9268, Validation Accuracy: 0.9553, Loss: 0.0500
Epoch   9 Batch 1780/2154 - Train Accuracy: 0.9469, Validation Accuracy: 0.9652, Loss: 0.0281
Epoch   9 Batch 1784/2154 - Train Accuracy: 0.9745, Validation Accuracy: 0.9531, Loss: 0.0219
Epoch   9 Batch 1788/2154 - Train Accuracy: 0.9398, Validation Accuracy: 0.9638, Loss: 0.0338
Epoch   9 Batch 1792/2154 - Train Accuracy: 0.9398, Validation Accuracy: 0.9709, Loss: 0.0332
Epoch   9 Batch 1796/2154 - Train Accuracy: 0.9554, Validati

Epoch   9 Batch 2108/2154 - Train Accuracy: 0.9546, Validation Accuracy: 0.9645, Loss: 0.0224
Epoch   9 Batch 2112/2154 - Train Accuracy: 0.9457, Validation Accuracy: 0.9574, Loss: 0.0311
Epoch   9 Batch 2116/2154 - Train Accuracy: 0.9539, Validation Accuracy: 0.9602, Loss: 0.0384
Epoch   9 Batch 2120/2154 - Train Accuracy: 0.9293, Validation Accuracy: 0.9439, Loss: 0.0463
Epoch   9 Batch 2124/2154 - Train Accuracy: 0.9523, Validation Accuracy: 0.9510, Loss: 0.0523
Epoch   9 Batch 2128/2154 - Train Accuracy: 0.9334, Validation Accuracy: 0.9489, Loss: 0.0344
Epoch   9 Batch 2132/2154 - Train Accuracy: 0.9406, Validation Accuracy: 0.9695, Loss: 0.0411
Epoch   9 Batch 2136/2154 - Train Accuracy: 0.9625, Validation Accuracy: 0.9688, Loss: 0.0510
Epoch   9 Batch 2140/2154 - Train Accuracy: 0.9353, Validation Accuracy: 0.9702, Loss: 0.0291
Epoch   9 Batch 2144/2154 - Train Accuracy: 0.9301, Validation Accuracy: 0.9361, Loss: 0.0422
Epoch   9 Batch 2148/2154 - Train Accuracy: 0.9430, Validati

### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [20]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [21]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [22]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    # TODO: Implement Function
    sentence = sentence.lower()
    words = []
    for word in sentence.split(' '):
        if not word:
            continue 
        else:
            try:
                words.append(vocab_to_int[word])
            except KeyError: 
                words.append(vocab_to_int['<UNK>'])                
        
    return words


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## Translate
This will translate `translate_sentence` from English to French.

In [30]:
translate_sentence = 'she is going to school .'


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('predictions:0')
    target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
    source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                         target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                         source_sequence_length: [len(translate_sentence)]*batch_size,
                                         keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in translate_logits]))
print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))


INFO:tensorflow:Restoring parameters from checkpoints/dev
Input
  Word Ids:      [8, 67, 133, 79, 53, 170]
  English Words: ['she', 'is', 'going', 'to', 'school', '.']

Prediction
  Word Ids:      [48, 166, 211, 313, 357, 67, 1]
  French Words: elle va à août prochain . <EOS>


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.